// @adonisjs/core/build/providers/edge_provider.js
import "../chunk-iKc69rpz.js";
import { t as main_exports } from "../main-DN2qEEg5.js";
import { pluginEdgeDumper } from "../modules/dumper/plugins/edge.js";
import edge from "edge.js";
var EdgeServiceProvider = class {
	constructor(app) {
		this.app = app;
		this.app.usingEdgeJS = true;
	}
	async boot() {
		const app = this.app;
		const qs = new main_exports.Qs(app.config.get("app.http.qs", {}));
		const router = await this.app.container.make("router");
		const dumper = await this.app.container.make("dumper");
		function edgeConfigResolver(key, defaultValue) {
			return app.config.get(key, defaultValue);
		}
		edgeConfigResolver.has = function(key) {
			return app.config.has(key);
		};
		function clientRoutes() {
			const routes = router.toJSON();
			return Object.keys(routes).reduce((result, domain) => {
				result[domain] = routes[domain].reduce((routesResult, route) => {
					if (!route.name) return routesResult;
					routesResult.push({
						domain: route.domain,
						methods: route.methods,
						pattern: route.pattern,
						tokens: route.tokens,
						name: route.name
					});
					return routesResult;
				}, []);
				return result;
			}, {});
		}
		edge.mount(app.viewsPath());
		edge.configure({ cache: app.inProduction });
		edge.global("route", function(...args) {
			return router.makeUrl(...args);
		});
		edge.global("signedRoute", function(...args) {
			return router.makeSignedUrl(...args);
		});
		edge.global("app", app);
		edge.global("config", edgeConfigResolver);
		edge.global("routes", function() {
			return clientRoutes();
		});
		edge.global("routesJSON", function() {
			return JSON.stringify(clientRoutes());
		});
		edge.global("urlFor", function(...args) {
			return router.urlBuilder.urlFor(...args);
		});
		edge.global("signedUrlFor", function(...args) {
			return router.urlBuilder.signedUrlFor(...args);
		});
		edge.global("qs", qs);
		edge.global("formAttributes", function(route, params, options) {
			const matchingRoute = router.findOrFail(route);
			options = options ?? {};
			let method = matchingRoute.methods[0].toUpperCase();
			const original = method;
			if (method === "HEAD") method = "GET";
			if (method !== "GET" && method !== "POST") {
				method = "POST";
				options = {
					...options,
					qs: {
						_method: original,
						...options.qs
					}
				};
			}
			const { action } = router.urlBuilder.urlFor.method(original, route, params, options).form;
			return {
				action,
				method
			};
		});
		main_exports.HttpContext.getter("view", function() {
			return edge.createRenderer().share({ request: this.request });
		}, true);
		main_exports.BriskRoute.macro("render", function(template, data) {
			function rendersTemplate({ view }) {
				return view.render(template, data);
			}
			Object.defineProperty(rendersTemplate, "listArgs", {
				value: template,
				writable: false
			});
			return this.setHandler(rendersTemplate);
		});
		edge.use(pluginEdgeDumper(dumper));
	}
};
export { EdgeServiceProvider as default };
