// @adonisjs/http-server/build/factories/main.js
import "../utils-BjSHKI3s.js";
import { _ as Qs, c as HttpRequest, i as HttpResponse, n as Server, r as HttpContext, s as Router, t as defineConfig } from "../define_config-D-kQXU0e.js";
import { t as createURL } from "../helpers-C_2HouOe.js";
import { parseRoute } from "../src/helpers.js";
import "../types-AUwURgIL.js";
import { Container } from "@adonisjs/fold";
import { Socket } from "node:net";
import proxyAddr from "proxy-addr";
import { safeStringify } from "@poppinss/utils/json";
import { AppFactory } from "@adonisjs/application/factories";
import { EncryptionFactory } from "@boringnode/encryption/factories";
import { randomUUID } from "node:crypto";
import { IncomingMessage, ServerResponse } from "node:http";
import { Logger } from "@adonisjs/logger";
import { Emitter } from "@adonisjs/events";
import { LoggerFactory } from "@adonisjs/logger/factories";
var QsParserFactory = class {
	#options = {
		parse: {
			depth: 5,
			parameterLimit: 1e3,
			allowSparse: false,
			arrayLimit: 20,
			comma: true
		},
		stringify: {
			encode: true,
			encodeValuesOnly: false,
			arrayFormat: "indices",
			skipNulls: false
		}
	};
	merge(options) {
		Object.assign(this.#options.parse, options.parse);
		Object.assign(this.#options.stringify, options.stringify);
		return this;
	}
	create() {
		return new Qs(this.#options);
	}
};
var RouterFactory = class {
	#parameters = {};
	#getApp() {
		return this.#parameters.app || new AppFactory().create(new URL("./app/", import.meta.url));
	}
	#createEncryption() {
		return this.#parameters.encryption || new EncryptionFactory().create();
	}
	merge(params) {
		Object.assign(this.#parameters, params);
		return this;
	}
	create() {
		return new Router(this.#getApp(), this.#createEncryption(), new QsParserFactory().create());
	}
};
var HttpRequestFactory = class {
	#parameters = {};
	#getConfig() {
		return {
			allowMethodSpoofing: false,
			trustProxy: proxyAddr.compile("loopback"),
			subdomainOffset: 2,
			generateRequestId: false,
			createRequestId() {
				return randomUUID();
			},
			...this.#parameters.config
		};
	}
	#createRequest() {
		const req = this.#parameters.req || new IncomingMessage(new Socket());
		if (this.#parameters.url) req.url = this.#parameters.url;
		if (this.#parameters.method) req.method = this.#parameters.method;
		return req;
	}
	#createResponse(req) {
		return this.#parameters.res || new ServerResponse(req);
	}
	#createEncryption() {
		return this.#parameters.encryption || new EncryptionFactory().create();
	}
	merge(params) {
		Object.assign(this.#parameters, params);
		return this;
	}
	create() {
		const req = this.#createRequest();
		return new HttpRequest(req, this.#createResponse(req), this.#createEncryption(), this.#getConfig(), new QsParserFactory().create());
	}
};
var HttpResponseFactory = class {
	#parameters = {};
	#getConfig() {
		return {
			etag: false,
			serializeJSON: safeStringify,
			jsonpCallbackName: "callback",
			cookie: {
				maxAge: 90,
				path: "/",
				httpOnly: true,
				sameSite: false,
				secure: false
			},
			...this.#parameters.config
		};
	}
	#createRequest() {
		return this.#parameters.req || new IncomingMessage(new Socket());
	}
	#createRouter() {
		return this.#parameters.router || new RouterFactory().create();
	}
	#createResponse(req) {
		return this.#parameters.res || new ServerResponse(req);
	}
	#createEncryption() {
		return this.#parameters.encryption || new EncryptionFactory().create();
	}
	merge(params) {
		Object.assign(this.#parameters, params);
		return this;
	}
	create() {
		const req = this.#createRequest();
		return new HttpResponse(req, this.#createResponse(req), this.#createEncryption(), this.#getConfig(), this.#createRouter(), new QsParserFactory().create());
	}
};
var ServerFactory = class {
	#parameters = {};
	#getEmitter() {
		return this.#parameters.emitter || new Emitter(this.#getApp());
	}
	#getLogger() {
		return this.#parameters.logger || new Logger({ enabled: false });
	}
	#getConfig() {
		return defineConfig(this.#parameters.config || {});
	}
	#getApp() {
		return this.#parameters.app || new AppFactory().create(new URL("./app/", import.meta.url));
	}
	#createEncryption() {
		return this.#parameters.encryption || new EncryptionFactory().create();
	}
	merge(params) {
		Object.assign(this.#parameters, params);
		return this;
	}
	create() {
		return new Server(this.#getApp(), this.#createEncryption(), this.#getEmitter(), this.#getLogger(), this.#getConfig());
	}
};
var HttpContextFactory = class {
	#parameters = {};
	#createRequest() {
		if (this.#parameters.request) return this.#parameters.request;
		let { url, method, route } = this.#parameters;
		if (route) url = createURL(route.pattern, parseRoute(route.pattern), new QsParserFactory().create().stringify, route.params);
		return new HttpRequestFactory().merge({
			url,
			method
		}).create();
	}
	#createResponse() {
		return this.#parameters.response || new HttpResponseFactory().create();
	}
	#createLogger() {
		return this.#parameters.logger || new LoggerFactory().create();
	}
	merge(params) {
		Object.assign(this.#parameters, params);
		return this;
	}
	create() {
		return new HttpContext(this.#createRequest(), this.#createResponse(), this.#createLogger(), new Container().createResolver());
	}
};
export { HttpContextFactory, QsParserFactory, HttpRequestFactory as RequestFactory, HttpResponseFactory as ResponseFactory, RouterFactory, ServerFactory };
