// @adonisjs/core/build/create_kernel-B7ILNhuP.js
import { d as Kernel, n as HelpCommand, t as FsLoader } from "./main-MBAMnmJb.js";
function createAceKernel(app, commandName) {
	const kernel = new Kernel(app);
	kernel.info.set("binary", "node ace");
	app.rcFile.commands.forEach((commandModule) => {
		kernel.addLoader(() => typeof commandModule === "function" ? commandModule() : app.import(commandModule));
	});
	const fsLoader = new FsLoader(app.commandsPath());
	kernel.addLoader({
		async getMetaData() {
			if (!commandName || !kernel.getCommand(commandName)) return fsLoader.getMetaData();
			return [];
		},
		getCommand(command) {
			return fsLoader.getCommand(command);
		}
	});
	kernel.defineFlag("ansi", {
		type: "boolean",
		showNegatedVariantInHelp: true,
		description: "Force enable or disable colorful output"
	});
	kernel.defineFlag("help", {
		type: "boolean",
		description: HelpCommand.description
	});
	kernel.on("ansi", (_, $kernel, parsed) => {
		if (parsed.flags.ansi === false) $kernel.ui.switchMode("silent");
		if (parsed.flags.ansi === true) $kernel.ui.switchMode("normal");
	});
	kernel.on("help", async (command, $kernel, parsed) => {
		parsed.args.unshift(command.commandName);
		await new HelpCommand($kernel, parsed, kernel.ui, kernel.prompt).exec();
		return $kernel.shortcircuit();
	});
	return kernel;
}
export { createAceKernel as t };
