// @adonisjs/core/build/define_config-0oHaj43l.js
import { t as debug_default } from "./debug-CGQmxzGt.js";
import { t as configProvider } from "./config_provider-FIAUgvae.js";
import { InvalidArgumentsException } from "./src/exceptions.js";
function defineConfig(config) {
	if (!config.list) throw new InvalidArgumentsException("Missing \"list\" property in encryption config");
	if (config.default && !config.list[config.default]) throw new InvalidArgumentsException(`Missing "list.${String(config.default)}" in encryption config. It is referenced by the "default" property`);
	return configProvider.create(async (app) => {
		debug_default("resolving encryption config");
		const encryptorsList = Object.keys(config.list);
		const encryptors = {};
		for (let encryptorName of encryptorsList) {
			const encryptor = config.list[encryptorName];
			if ("resolver" in encryptor) encryptors[encryptorName] = await encryptor.resolver(app);
			else encryptors[encryptorName] = encryptor;
		}
		return {
			default: config.default,
			list: encryptors
		};
	});
}
const drivers = {
	chacha20: (config) => {
		return configProvider.create(async () => {
			const { ChaCha20Poly1305 } = await import("./modules/encryption/drivers/chacha20_poly1305.js");
			debug_default("configuring chacha20 encryption driver");
			return {
				driver: (key) => new ChaCha20Poly1305({
					id: config.id,
					key
				}),
				keys: config.keys.filter((key) => !!key)
			};
		});
	},
	aes256cbc: (config) => {
		return configProvider.create(async () => {
			const { AES256CBC } = await import("./modules/encryption/drivers/aes_256_cbc.js");
			debug_default("configuring aes256cbc encryption driver");
			return {
				driver: (key) => new AES256CBC({
					id: config.id,
					key
				}),
				keys: config.keys.filter((key) => !!key)
			};
		});
	},
	aes256gcm: (config) => {
		return configProvider.create(async () => {
			const { AES256GCM } = await import("./modules/encryption/drivers/aes_256_gcm.js");
			debug_default("configuring aes256gcm encryption driver");
			return {
				driver: (key) => new AES256GCM({
					id: config.id,
					key
				}),
				keys: config.keys.filter((key) => !!key)
			};
		});
	},
	aessiv: (config) => {
		return configProvider.create(async () => {
			const { AESSIV } = await import("./modules/encryption/drivers/aes_siv.js");
			debug_default("configuring aessiv encryption driver");
			return {
				driver: (key) => new AESSIV({
					id: config.id,
					key
				}),
				keys: [config.key].filter((key) => !!key)
			};
		});
	},
	legacy: (config) => {
		return configProvider.create(async () => {
			const { Legacy } = await import("./modules/encryption/drivers/legacy.js");
			debug_default("configuring legacy encryption driver");
			return {
				driver: (key) => new Legacy({ key }),
				keys: config.keys.filter((key) => !!key)
			};
		});
	}
};
export { drivers as n, defineConfig as t };
