// @adonisjs/core/build/dumper-BBgqFX5a.js
import { t as __exportAll } from "./chunk-iKc69rpz.js";
import { inspect } from "node:util";
import { Exception } from "@poppinss/utils/exception";
import useColors from "@poppinss/colors";
import { dump } from "@poppinss/dumper/console";
import { createScript, createStyleSheet, dump as dump$1 } from "@poppinss/dumper/html";
import { parse } from "error-stack-parser-es";
var errors_exports = /* @__PURE__ */ __exportAll({ E_DUMP_DIE_EXCEPTION: () => E_DUMP_DIE_EXCEPTION });
var DumpDieException = class extends Exception {
	static status = 500;
	static code = "E_DUMP_DIE_EXCEPTION";
	#dumper;
	#traceSourceIndex = 1;
	value;
	constructor(value, dumper) {
		super("Dump and Die exception");
		this.#dumper = dumper;
		this.value = value;
	}
	#getErrorSource() {
		if (this.fileName && this.lineNumber) return {
			location: this.fileName,
			line: this.lineNumber
		};
		const source = parse(this)[this.#traceSourceIndex];
		if (!source.fileName || !source.lineNumber) return;
		return {
			location: source.fileName,
			line: source.lineNumber
		};
	}
	setTraceSourceIndex(index) {
		this.#traceSourceIndex = index;
		return this;
	}
	report() {}
	async handle(error, ctx) {
		const source = this.#getErrorSource();
		const cspNonce = "nonce" in ctx.response ? ctx.response.nonce : void 0;
		ctx.response.status(500).send(`<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width">${this.#dumper.getHeadElements(cspNonce)}</head><body>${this.#dumper.dumpToHtml(error.value, {
			cspNonce,
			source,
			title: "DUMP DIE"
		})}</body></html>`);
	}
	async render(error, kernel) {
		const source = this.#getErrorSource();
		kernel.ui.logger.log(this.#dumper.dumpToAnsi(error.value, {
			source,
			title: "DUMP DIE"
		}));
	}
	[inspect.custom]() {
		const source = this.#getErrorSource();
		return this.#dumper.dumpToAnsi(this.value, {
			source,
			title: "DUMP DIE"
		});
	}
};
const E_DUMP_DIE_EXCEPTION = DumpDieException;
const colors = useColors.ansi();
const DUMP_TITLE_STYLES = `
.adonisjs-dump-header {
  font-family: JetBrains Mono, monaspace argon, Menlo, Monaco, Consolas, monospace;
  background: #ff1639;
  border-radius: 4px;
  color: #fff;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  padding: 0.4rem 1.2rem;
  font-size: 1em;
  display: flex;
  justify-content: space-between;
}
.adonisjs-dump-header .adonisjs-dump-header-title {
  font-weight: bold;
  text-transform: uppercase;
}
.adonisjs-dump-header .adonisjs-dump-header-source {
  font-weight: bold;
  color: inherit;
  text-decoration: underline;
}
.dumper-dump pre {
  border-radius: 4px;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
}`;
const IDE = process.env.ADONIS_IDE ?? process.env.EDITOR ?? "";
var Dumper = class {
	#app;
	#htmlConfig = {};
	#consoleConfig = { collapse: ["DateTime", "Date"] };
	#editors = {
		textmate: "txmt://open?url=file://%f&line=%l",
		macvim: "mvim://open?url=file://%f&line=%l",
		emacs: "emacs://open?url=file://%f&line=%l",
		sublime: "subl://open?url=file://%f&line=%l",
		phpstorm: "phpstorm://open?file=%f&line=%l",
		atom: "atom://core/open/file?filename=%f&line=%l",
		vscode: "vscode://file/%f:%l"
	};
	constructor(app) {
		this.#app = app;
	}
	#getEditorLink(source) {
		const editorURL = this.#editors[IDE] || IDE;
		if (!editorURL || !source) return;
		return {
			href: editorURL.replace("%f", source.location).replace("%l", String(source.line)),
			text: `${this.#app.relativePath(source.location)}:${source.line}`
		};
	}
	configureHtmlOutput(config) {
		this.#htmlConfig = config;
		return this;
	}
	configureAnsiOutput(config) {
		this.#consoleConfig = config;
		return this;
	}
	getHeadElements(cspNonce) {
		return `<style id="dumper-styles">` + createStyleSheet() + DUMP_TITLE_STYLES + `</style><script id="dumper-script"${cspNonce ? ` nonce="${cspNonce}"` : ""}>` + createScript() + "<\/script>";
	}
	dumpToHtml(value, options = {}) {
		const link = this.#getEditorLink(options.source) ?? null;
		return `<div class="adonisjs-dump-header"><span class="adonisjs-dump-header-title">${options.title || "DUMP"}</span>` + (link ? `<a href="${link.href}" class="adonisjs-dump-header-source">${link.text}</a>` : "") + "</div>" + dump$1(value, {
			cspNonce: options.cspNonce,
			...this.#htmlConfig
		});
	}
	dumpToAnsi(value, options = {}) {
		const columns = process.stdout.columns;
		const link = `${this.#getEditorLink(options.source)?.text ?? ""} `;
		const title = ` ${options.title || "DUMP"}`;
		const whiteSpaceLength = columns ? columns - link.length - title.length - 4 : 2;
		const whiteSpace = new Array(whiteSpaceLength <= 0 ? 2 : whiteSpaceLength).join(" ");
		return `${colors.bgRed().bold(`${title}${whiteSpace}${link}`)}\n${dump(value, this.#consoleConfig)}`;
	}
	dd(value, traceSourceIndex = 1) {
		const error = new E_DUMP_DIE_EXCEPTION(value, this);
		error.setTraceSourceIndex(traceSourceIndex);
		throw error;
	}
};
export { errors_exports as n, Dumper as t };
