// @adonisjs/core/build/main-DN2qEEg5.js
import { n as __reExport, t as __exportAll } from "./chunk-iKc69rpz.js";
import "@adonisjs/bodyparser";
var RequestValidator = class RequestValidator {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	static errorReporter;
	static messagesProvider;
	#requestData() {
		return {
			...this.#ctx.request.all(),
			params: this.#ctx.request.params(),
			headers: this.#ctx.request.headers(),
			cookies: this.#ctx.request.cookiesList()
		};
	}
	#processValidatorOptions(options) {
		const validatorOptions = options || {};
		if (RequestValidator.errorReporter && !validatorOptions.errorReporter) {
			const errorReporter = RequestValidator.errorReporter(this.#ctx);
			validatorOptions.errorReporter = () => errorReporter;
		}
		if (RequestValidator.messagesProvider && !validatorOptions.messagesProvider) validatorOptions.messagesProvider = RequestValidator.messagesProvider(this.#ctx);
		return validatorOptions;
	}
	validateUsing(validator, ...[options]) {
		const validatorOptions = this.#processValidatorOptions(options);
		const data = validatorOptions.data || this.#requestData();
		return validator.validate(data, validatorOptions);
	}
	async tryValidateUsing(validator, ...[options]) {
		const validatorOptions = this.#processValidatorOptions(options);
		const data = validatorOptions.data || this.#requestData();
		return validator.tryValidate(data, validatorOptions);
	}
};
var main_exports = /* @__PURE__ */ __exportAll({ RequestValidator: () => RequestValidator });
import * as import__adonisjs_http_server from "@adonisjs/http-server";
__reExport(main_exports, import__adonisjs_http_server);
export { RequestValidator as n, main_exports as t };
