// @adonisjs/core/build/main-MBAMnmJb.js
import { BaseCommand, FsLoader, HelpCommand as HelpCommand$1, IndexGenerator as IndexGenerator$1, Kernel, ListCommand, ListLoader as ListLoader$1, Parser, args as args$1, cliHelpers, errors as errors$1, flags, tracingChannels } from "@adonisjs/ace";
var BaseCommand$1 = class extends BaseCommand {
	static options = {};
	get staysAlive() {
		return this.constructor.options.staysAlive;
	}
	get startApp() {
		return this.constructor.options.startApp;
	}
	constructor(app, kernel, parsed, ui, prompt) {
		super(kernel, parsed, ui, prompt);
		this.app = app;
		this.kernel = kernel;
	}
	async createCodemods() {
		const { Codemods } = await import("./modules/ace/codemods.js");
		const codemods = new Codemods(this.app, this.logger);
		codemods.on("error", () => {
			this.exitCode = 1;
		});
		return codemods;
	}
	async exec() {
		this.hydrate();
		try {
			this.prepare && await this.app.container.call(this, "prepare");
			this.interact && await this.app.container.call(this, "interact");
			const result = await this.app.container.call(this, "run");
			this.result = this.result === void 0 ? result : this.result;
			this.exitCode = this.exitCode ?? 0;
		} catch (error) {
			this.error = error;
			this.exitCode = this.exitCode ?? 1;
		}
		let errorHandled = this.completed ? await this.app.container.call(this, "completed") : false;
		if (this.error && !errorHandled) await this.kernel.errorHandler.render(this.error, this.kernel);
		return this.result;
	}
	async terminate() {
		if (this.kernel.getMainCommand() === this) await this.app.terminate();
	}
};
var ListCommand$1 = class extends ListCommand {
	static options = {};
	get staysAlive() {
		return this.constructor.options.staysAlive;
	}
	get startApp() {
		return this.constructor.options.startApp;
	}
	constructor(app, kernel, parsed, ui, prompt) {
		super(kernel, parsed, ui, prompt);
		this.app = app;
		this.kernel = kernel;
	}
	async createCodemods() {
		const { Codemods } = await import("./modules/ace/codemods.js");
		return new Codemods(this.app, this.logger);
	}
	async terminate() {
		if (this.kernel.getMainCommand() === this) await this.app.terminate();
	}
};
var Kernel$1 = class extends Kernel {
	constructor(app) {
		super(ListCommand$1, {
			create: async (command, parsedOutput, $kernel) => {
				return app.container.make(command, [
					app,
					$kernel,
					parsedOutput,
					$kernel.ui,
					$kernel.prompt
				]);
			},
			run: (command) => command.exec()
		});
		this.app = app;
	}
};
export { Parser as a, errors$1 as c, Kernel$1 as d, BaseCommand$1 as f, ListLoader$1 as i, flags as l, HelpCommand$1 as n, args$1 as o, ListCommand$1 as p, IndexGenerator$1 as r, cliHelpers as s, FsLoader as t, tracingChannels as u };
