// @adonisjs/core/build/main-DkNgvceD.js
import { n as __reExport, t as __exportAll } from "./chunk-iKc69rpz.js";
import { t as debug_default } from "./debug-CGQmxzGt.js";
import { t as configProvider } from "./config_provider-FIAUgvae.js";
import { InvalidArgumentsException } from "@poppinss/utils/exception";
function defineConfig(config) {
	if (!config.list) throw new InvalidArgumentsException("Missing \"list\" property in hash config");
	if (config.default && !config.list[config.default]) throw new InvalidArgumentsException(`Missing "list.${String(config.default)}" in hash config. It is referenced by the "default" property`);
	return configProvider.create(async (app) => {
		debug_default("resolving hash config");
		const hashersList = Object.keys(config.list);
		const hashers = {};
		for (let hasherName of hashersList) {
			const hasher = config.list[hasherName];
			if (typeof hasher === "function") hashers[hasherName] = hasher;
			else hashers[hasherName] = await hasher.resolver(app);
		}
		return {
			default: config.default,
			list: hashers
		};
	});
}
const drivers = {
	argon2: (config) => {
		return configProvider.create(async () => {
			const { Argon } = await import("./modules/hash/drivers/argon.js");
			debug_default("configuring argon driver");
			return () => new Argon(config);
		});
	},
	bcrypt: (config) => {
		return configProvider.create(async () => {
			const { Bcrypt } = await import("./modules/hash/drivers/bcrypt.js");
			debug_default("configuring bcrypt driver");
			return () => new Bcrypt(config);
		});
	},
	scrypt: (config) => {
		return configProvider.create(async () => {
			const { Scrypt } = await import("./modules/hash/drivers/scrypt.js");
			debug_default("configuring scrypt driver");
			return () => new Scrypt(config);
		});
	}
};
var main_exports = /* @__PURE__ */ __exportAll({
	defineConfig: () => defineConfig,
	drivers: () => drivers
});
import * as import__adonisjs_hash from "@adonisjs/hash";
__reExport(main_exports, import__adonisjs_hash);
export { defineConfig as n, drivers as r, main_exports as t };
