// @adonisjs/core/build/main--nXd7T-C.js
import { t as debug_default } from "./debug-CGQmxzGt.js";
import { t as main_exports } from "./main-DN2qEEg5.js";
import { Encryption } from "./modules/encryption/main.js";
import { IncomingMessage, ServerResponse, createServer } from "node:http";
import { Socket } from "node:net";
import Macroable from "@poppinss/macroable";
var HttpServerUtils = class {
	#utils;
	constructor(utils) {
		this.#utils = utils;
	}
	#listen(nodeHttpServer) {
		return new Promise((resolve, reject) => {
			const host = process.env.HOST || "0.0.0.0";
			const port = Number(process.env.PORT || "3333");
			nodeHttpServer.listen(port, host);
			nodeHttpServer.once("listening", () => {
				debug_default("listening to utils http server, host :%s, port: %s", host, port);
				resolve({
					port,
					host
				});
			});
			nodeHttpServer.once("error", (error) => {
				reject(error);
			});
		});
	}
	async start(serverCallback) {
		const createHTTPServer = serverCallback || createServer;
		const server = await this.#utils.app.container.make("server");
		await server.boot();
		const httpServer = createHTTPServer(server.handle.bind(server));
		server.setNodeServer(httpServer);
		await this.#listen(httpServer);
		return () => {
			return new Promise((resolve, reject) => {
				httpServer.close((error) => {
					if (error) reject(error);
					else resolve();
				});
			});
		};
	}
};
var TestUtils = class extends Macroable {
	#booted = false;
	get isBooted() {
		return this.#booted;
	}
	constructor(app) {
		super();
		this.app = app;
	}
	async boot() {
		if (!this.isBooted) {
			this.#booted = true;
			this.cookies = new main_exports.CookieClient(await this.app.container.make(Encryption));
		}
	}
	httpServer() {
		return new HttpServerUtils(this);
	}
	async createHttpContext(options = {}) {
		const req = options.req || new IncomingMessage(new Socket());
		const res = options.res || new ServerResponse(req);
		const server = await this.app.container.make("server");
		const request = server.createRequest(req, res);
		const response = server.createResponse(req, res);
		return server.createHttpContext(request, response, this.app.container.createResolver());
	}
};
export { TestUtils as t };
