// @adonisjs/core/build/utils-rRkbAPnP.js
async function importAssembler(app) {
	try {
		return await app.import("@adonisjs/assembler");
	} catch {}
}
async function importTypeScript(app) {
	try {
		return await app.importDefault("typescript");
	} catch {}
}
async function outputTransformerDataObjects(transformersList, buffer, withSharedProps, inertiaMiddlewareImportPath = "#middleware/inertia_middleware") {
	const importsBuffer = buffer.create();
	importsBuffer.write(`/// <reference path="./manifest.d.ts" />`);
	importsBuffer.write(`import type { InferData, InferVariants } from '@adonisjs/core/types/transformers'`);
	if (withSharedProps) importsBuffer.write(`import type { InferSharedProps } from '@adonisjs/inertia/types'`);
	buffer.writeLine(importsBuffer);
	buffer.write("export namespace Data {").indent();
	function generateNamespaceTree(input, parents) {
		Object.keys(input).forEach((key) => {
			const value = input[key];
			if (typeof value === "string") {
				const importName = `${parents.join("")}${key}Transformer`;
				importsBuffer.write(`import type ${importName} from '${value}'`);
				buffer.write(`export type ${key} = InferData<${importName}>`);
				buffer.write(`export namespace ${key} {`).indent();
				buffer.write(`export type Variants = InferVariants<${importName}>`);
				buffer.dedent().write("}");
			} else {
				buffer.write(`export namespace ${key} {`).indent();
				generateNamespaceTree(value, [...parents, key]);
				buffer.dedent().write(`}`);
			}
		});
	}
	generateNamespaceTree(transformersList, []);
	if (withSharedProps) {
		importsBuffer.write(`import type InertiaMiddleware from '${inertiaMiddlewareImportPath}'`);
		buffer.write("export type SharedProps = InferSharedProps<InertiaMiddleware>");
	}
	buffer.dedent().write("}");
}
export { importTypeScript as n, outputTransformerDataObjects as r, importAssembler as t };
