// @boringnode/encryption/build/chunk-5QCRCQLG.js
import {
  MessageVerifier
} from "./chunk-SMDTGIFR.js";

// src/encryption.ts
var Encryption = class {
  #drivers;
  #verifier;
  constructor(config) {
    this.#drivers = config.keys.map((key) => config.driver(key));
    this.#verifier = new MessageVerifier(config.keys);
  }
  encrypt(payload, expiresInOrOptions, purpose) {
    let expiresIn;
    let actualPurpose;
    if (typeof expiresInOrOptions === "object" && expiresInOrOptions !== null) {
      expiresIn = expiresInOrOptions.expiresIn;
      actualPurpose = expiresInOrOptions.purpose;
    } else {
      expiresIn = expiresInOrOptions;
      actualPurpose = purpose;
    }
    return this.#drivers[0].encrypt(payload, expiresIn, actualPurpose);
  }
  /**
   * Decrypt a value by trying all keys
   */
  decrypt(value, purpose) {
    for (const driver of this.#drivers) {
      const result = driver.decrypt(value, purpose);
      if (result !== null) {
        return result;
      }
    }
    return null;
  }
  blindIndex(payload, purpose) {
    return this.#drivers[0].blindIndex(payload, purpose);
  }
  blindIndexes(payload, purpose) {
    const indexes = /* @__PURE__ */ new Set();
    for (const driver of this.#drivers) {
      for (const index of driver.blindIndexes(payload, purpose)) {
        indexes.add(index);
      }
    }
    return [...indexes];
  }
  /**
   * Get the message verifier instance
   */
  getMessageVerifier() {
    return this.#verifier;
  }
};

export {
  Encryption
};
/*
 * @boringnode/encryption
 *
 * @license MIT
 * @copyright Boring Node
 */
