// @boringnode/encryption/build/chunk-D5BCA36S.js
import {
  BaseDriver,
  E_MISSING_ENCRYPTER_ID
} from "./chunk-7R6WELCL.js";
import {
  base64UrlDecode,
  base64UrlEncode
} from "./chunk-ACZGKXGJ.js";

// src/drivers/chacha20_poly1305.ts
import { createCipheriv, createDecipheriv, randomBytes } from "crypto";
import { MessageBuilder } from "@poppinss/utils";
function chacha20poly1305(config) {
  return {
    driver: (key) => new ChaCha20Poly1305({ id: config.id, key }),
    keys: config.keys
  };
}
var ChaCha20Poly1305 = class extends BaseDriver {
  #config;
  constructor(config) {
    super(config);
    this.#config = config;
    if (typeof config.id !== "string") {
      throw new E_MISSING_ENCRYPTER_ID();
    }
  }
  encrypt(payload, expiresInOrOptions, purpose) {
    let expiresIn;
    let actualPurpose;
    if (typeof expiresInOrOptions === "object" && expiresInOrOptions !== null) {
      expiresIn = expiresInOrOptions.expiresIn;
      actualPurpose = expiresInOrOptions.purpose;
    } else {
      expiresIn = expiresInOrOptions;
      actualPurpose = purpose;
    }
    const iv = randomBytes(12);
    const cipher = createCipheriv("chacha20-poly1305", this.cryptoKey, iv, {
      authTagLength: 16
    });
    if (actualPurpose) {
      cipher.setAAD(Buffer.from(actualPurpose), {
        plaintextLength: Buffer.byteLength(actualPurpose)
      });
    }
    const plainText = new MessageBuilder().build(payload, expiresIn);
    const cipherText = Buffer.concat([cipher.update(plainText), cipher.final()]);
    const tag = cipher.getAuthTag();
    return this.computeReturns([
      this.#config.id,
      base64UrlEncode(cipherText),
      base64UrlEncode(iv),
      base64UrlEncode(tag)
    ]);
  }
  /**
   * Decrypt value and verify it against a purpose
   */
  decrypt(value, purpose) {
    if (typeof value !== "string") {
      return null;
    }
    const [id, cipherEncoded, ivEncoded, tagEncoded] = value.split(this.separator);
    if (!id || !cipherEncoded || !ivEncoded || !tagEncoded) {
      return null;
    }
    if (id !== this.#config.id) {
      return null;
    }
    const cipherText = base64UrlDecode(cipherEncoded);
    if (!cipherText) {
      return null;
    }
    const iv = base64UrlDecode(ivEncoded);
    if (!iv) {
      return null;
    }
    const tag = base64UrlDecode(tagEncoded);
    if (!tag) {
      return null;
    }
    try {
      const decipher = createDecipheriv("chacha20-poly1305", this.cryptoKey, iv, {
        authTagLength: 16
      });
      if (purpose) {
        decipher.setAAD(Buffer.from(purpose), { plaintextLength: Buffer.byteLength(purpose) });
      }
      decipher.setAuthTag(tag);
      const plain = Buffer.concat([decipher.update(cipherText), decipher.final()]);
      return new MessageBuilder().verify(plain);
    } catch {
      return null;
    }
  }
};

export {
  chacha20poly1305,
  ChaCha20Poly1305
};
/*
 * @boringnode/encryption
 *
 * @license MIT
 * @copyright Boring Node
 */
