// parse-imports/dist/index.js
import module, { createRequire } from "node:module";
import { init, parse } from "es-module-lexer";
import assert from "node:assert";
import { removeSlashes } from "slashes";
import { dirname } from "node:path";

//#region src/parse-import-clause/skip.js
const separatorRegex = /^(?:\s+|,)$/u;
const skipSeparators = (imported, i) => {
	while (i < imported.length && separatorRegex.test(imported[i])) i++;
	return i;
};
const skipNonSeparators = (imported, i) => {
	while (i < imported.length && !separatorRegex.test(imported[i])) i++;
	return i;
};

//#endregion
//#region src/parse-import-clause/parse-default-import.js
const parseDefaultImport = (importClauseString, i) => {
	const startIndex = i;
	i = skipNonSeparators(importClauseString, i);
	return {
		defaultImport: importClauseString.slice(startIndex, i),
		i
	};
};
var parse_default_import_default = parseDefaultImport;

//#endregion
//#region src/parse-import-clause/parse-named-imports.js
const parseNamedImports = (importClauseString, i) => {
	const startIndex = ++i;
	while (i < importClauseString.length && importClauseString[i] !== `}`) i++;
	const namedImports = importClauseString.slice(startIndex, i++).split(`,`).map((namedImport) => {
		namedImport = namedImport.trim();
		if (namedImport.includes(` `)) {
			const components = namedImport.split(` `);
			return {
				specifier: components[0],
				binding: components.at(-1)
			};
		}
		return {
			specifier: namedImport,
			binding: namedImport
		};
	}).filter(({ specifier }) => specifier.length > 0);
	return {
		namedImports,
		i
	};
};
var parse_named_imports_default = parseNamedImports;

//#endregion
//#region src/parse-import-clause/parse-namespace-import.js
const parseNamespaceImport = (importClauseString, i) => {
	i++;
	i = skipSeparators(importClauseString, i);
	i += `as`.length;
	i = skipSeparators(importClauseString, i);
	const startIndex = i;
	i = skipNonSeparators(importClauseString, i);
	return {
		namespaceImport: importClauseString.slice(startIndex, i),
		i
	};
};
var parse_namespace_import_default = parseNamespaceImport;

//#endregion
//#region src/parse-import-clause/index.js
const parseImportClause = (importClauseString) => {
	let defaultImport;
	let namespaceImport;
	const namedImports = [];
	for (let i = 0; i < importClauseString.length; i++) {
		if (separatorRegex.test(importClauseString[i])) continue;
		if (importClauseString[i] === `{`) {
			let newNamedImports;
			({namedImports: newNamedImports, i} = parse_named_imports_default(importClauseString, i));
			namedImports.push(...newNamedImports);
		} else if (importClauseString[i] === `*`) ({namespaceImport, i} = parse_namespace_import_default(importClauseString, i));
		else ({defaultImport, i} = parse_default_import_default(importClauseString, i));
	}
	return {
		default: defaultImport,
		namespace: namespaceImport,
		named: namedImports
	};
};
var parse_import_clause_default = parseImportClause;

//#endregion
//#region src/parse-module-specifier/is-constant-string-literal.js
const isConstantStringLiteral = (stringLiteral) => {
	const quote = [
		`'`,
		`"`,
		`\``
	].find((quoteCandidate) => stringLiteral.startsWith(quoteCandidate) && stringLiteral.endsWith(quoteCandidate));
	if (quote == null) return false;
	for (let i = 1; i < stringLiteral.length - 1; i++) {
		if (stringLiteral[i] === quote && stringLiteral[i - 1] !== `\\`) return false;
		if (quote === `\`` && stringLiteral.slice(i, i + 2) === `\${` && stringLiteral[i - 1] !== `\\`) return false;
	}
	return true;
};
var is_constant_string_literal_default = isConstantStringLiteral;

//#endregion
//#region src/parse-module-specifier/parse-type.js
const builtinModules = new Set(module.builtinModules);
const parseType = (moduleSpecifier) => {
	if (moduleSpecifier.length === 0) return `invalid`;
	if (moduleSpecifier.startsWith(`/`)) return `absolute`;
	if (moduleSpecifier.startsWith(`.`)) return `relative`;
	if (builtinModules.has(moduleSpecifier)) return `builtin`;
	return `package`;
};
var parse_type_default = parseType;

//#endregion
//#region src/parse-module-specifier/resolve.js
const require = createRequire(import.meta.url);
const resolve = (from, to) => {
	try {
		return require.resolve(to, { paths: [dirname(from)] });
	} catch {
		return void 0;
	}
};
var resolve_default = resolve;

//#endregion
//#region src/parse-module-specifier/index.js
const parseModuleSpecifier = (moduleSpecifierString, { isDynamicImport, resolveFrom }) => {
	assert(isDynamicImport || is_constant_string_literal_default(moduleSpecifierString));
	const { isConstant, value } = !isDynamicImport || is_constant_string_literal_default(moduleSpecifierString) ? {
		isConstant: true,
		value: removeSlashes(moduleSpecifierString.slice(1, -1))
	} : {
		isConstant: false,
		value: void 0
	};
	return {
		type: isConstant ? parse_type_default(value) : `unknown`,
		isConstant,
		code: moduleSpecifierString,
		value,
		resolved: typeof resolveFrom === `string` && isConstant ? resolve_default(resolveFrom, value) : void 0
	};
};
var parse_module_specifier_default = parseModuleSpecifier;

//#endregion
//#region src/index.js
const wasmLoadPromise = init;
const parseImports = async (code, options) => {
	await wasmLoadPromise;
	return parseImportsSync(code, options);
};
const parseImportsSync = (code, { resolveFrom } = {}) => {
	const result = parse(code, resolveFrom == null ? void 0 : resolveFrom);
	if (!Array.isArray(result)) throw new TypeError(`Expected WASM to be loaded before calling parseImportsSync`);
	const [imports] = result;
	return { *[Symbol.iterator]() {
		for (let { d: dynamicImportStartIndex, ss: statementStartIndex, s: moduleSpecifierStartIndex, e: moduleSpecifierEndIndexExclusive } of imports) {
			const isImportMeta = dynamicImportStartIndex === -2;
			if (isImportMeta) continue;
			const isDynamicImport = dynamicImportStartIndex > -1;
			if (!isDynamicImport) {
				moduleSpecifierStartIndex--;
				moduleSpecifierEndIndexExclusive++;
			}
			const moduleSpecifierString = code.slice(moduleSpecifierStartIndex, moduleSpecifierEndIndexExclusive);
			const moduleSpecifier = {
				startIndex: moduleSpecifierStartIndex,
				endIndex: moduleSpecifierEndIndexExclusive,
				...parse_module_specifier_default(moduleSpecifierString, {
					isDynamicImport,
					resolveFrom
				})
			};
			let importClause;
			if (!isDynamicImport) {
				let importClauseString = code.slice(statementStartIndex + `import`.length, moduleSpecifierStartIndex).trim();
				if (importClauseString.endsWith(`from`)) importClauseString = importClauseString.slice(0, Math.max(0, importClauseString.length - `from`.length));
				importClause = parse_import_clause_default(importClauseString);
			}
			yield {
				startIndex: statementStartIndex,
				endIndex: isDynamicImport ? moduleSpecifierEndIndexExclusive + 1 : moduleSpecifierEndIndexExclusive,
				isDynamicImport,
				moduleSpecifier,
				importClause
			};
		}
	} };
};

//#endregion
export { parseImports, parseImportsSync, wasmLoadPromise };
