// @adonisjs/core/build/factories/core/ignitor.js
import "../../chunk-iKc69rpz.js";
import "../../debug-CGQmxzGt.js";
import { t as Ignitor } from "../../main-Cxz0TyIw.js";
import "../../config_provider-FIAUgvae.js";
import { t as logger_exports } from "../../modules/logger.js";
import { t as main_exports } from "../../main-DN2qEEg5.js";
import { n as drivers, t as defineConfig } from "../../define_config-0oHaj43l.js";
import { n as defineConfig$1, r as drivers$1 } from "../../main-DkNgvceD.js";
import { t as main_exports$1 } from "../../modules/bodyparser/main.js";
var IgnitorFactory = class {
	#preloadActions = [];
	#parameters = {};
	#loadCoreProviders = false;
	preload(action) {
		this.#preloadActions.push(action);
		return this;
	}
	#mergeCoreProviders(providers) {
		return [
			() => import("@adonisjs/core/providers/app_provider"),
			() => import("@adonisjs/core/providers/hash_provider"),
			() => import("@adonisjs/core/providers/repl_provider")
		].concat(providers || []);
	}
	merge(params) {
		if (params.config) this.#parameters.config = Object.assign(this.#parameters.config || {}, params.config);
		if (params.rcFileContents) this.#parameters.rcFileContents = Object.assign(this.#parameters.rcFileContents || {}, params.rcFileContents);
		return this;
	}
	withCoreProviders() {
		this.#loadCoreProviders = true;
		return this;
	}
	withCoreConfig() {
		this.merge({ config: {
			app: {
				appUrl: "http://localhost:3333",
				http: (0, main_exports.defineConfig)({})
			},
			validator: {},
			encryption: defineConfig({
				default: "gcm",
				list: { gcm: drivers.aes256gcm({
					id: "gcm",
					keys: ["averylongrandomsecretkey"]
				}) }
			}),
			bodyparser: (0, main_exports$1.defineConfig)({}),
			hash: defineConfig$1({
				default: "scrypt",
				list: { scrypt: drivers$1.scrypt({}) }
			}),
			logger: (0, logger_exports.defineConfig)({
				default: "app",
				loggers: { app: {} }
			})
		} });
		return this;
	}
	create(appRoot, options) {
		return new Ignitor(appRoot, options).tap((app) => {
			app.booted(async () => {
				for (let action of this.#preloadActions) await action(app);
			});
			if (this.#loadCoreProviders) {
				this.#parameters.rcFileContents = this.#parameters.rcFileContents || {};
				this.#parameters.rcFileContents.providers = this.#mergeCoreProviders(this.#parameters.rcFileContents.providers);
			}
			this.#parameters.rcFileContents && app.rcContents(this.#parameters.rcFileContents);
			this.#parameters.config && app.useConfig(this.#parameters.config);
		});
	}
};
export { IgnitorFactory };
