// @adonisjs/logger/build/logger-Btq_FVdY.js
import abstractLogging from "abstract-logging";
import { destination, multistream, pino, stdSerializers, stdTimeFunctions, transport, version } from "pino";
const TimestampFormatters = {
	iso: stdTimeFunctions.isoTime,
	epoch: stdTimeFunctions.epochTime,
	unix: stdTimeFunctions.unixTime
};
function createPino(options) {
	const { desination, transport: configTransport, timestamp, ...rest } = options;
	const pinoOptions = desination ? Object.assign({}, rest) : Object.assign({ transport: configTransport }, rest);
	if (typeof timestamp === "string" && TimestampFormatters[timestamp]) pinoOptions.timestamp = TimestampFormatters[timestamp];
	else if (typeof timestamp === "boolean" || typeof timestamp === "function") pinoOptions.timestamp = timestamp;
	return desination ? pino(pinoOptions, desination) : pino(pinoOptions);
}
const levels = {
	labels: {
		10: "trace",
		20: "debug",
		30: "info",
		40: "warn",
		50: "error",
		60: "fatal"
	},
	values: {
		fatal: 60,
		error: 50,
		warn: 40,
		info: 30,
		debug: 20,
		trace: 10
	}
};
var Logger = class Logger {
	pino;
	constructor(config, pino) {
		this.config = config;
		if (!this.config.enabled) this.pino = abstractLogging;
		else this.pino = pino || createPino(this.config);
	}
	get isEnabled() {
		return !!this.config.enabled;
	}
	get levels() {
		if (!this.isEnabled) return levels;
		return this.pino.levels;
	}
	get level() {
		if (!this.isEnabled) return this.config.level || "info";
		return this.pino.level;
	}
	set level(level) {
		if (!this.isEnabled) {
			this.config.level = level;
			return;
		}
		this.pino.level = level;
	}
	get levelNumber() {
		if (!this.isEnabled) return levels.values[this.config.level || "info"];
		return this.pino.levelVal;
	}
	get pinoVersion() {
		return version;
	}
	get version() {
		return version;
	}
	isLevelEnabled(level) {
		if (!this.isEnabled) return false;
		return this.pino.isLevelEnabled(level);
	}
	ifLevelEnabled(level, callback) {
		if (this.isLevelEnabled(level)) return callback(this);
	}
	log(level, mergingObject, message, ...values) {
		if (!this.isEnabled) return;
		this.pino[level](mergingObject, message, ...values);
	}
	trace(mergingObject, message, ...values) {
		this.log("trace", mergingObject, message, ...values);
	}
	debug(mergingObject, message, ...values) {
		this.log("debug", mergingObject, message, ...values);
	}
	info(mergingObject, message, ...values) {
		this.log("info", mergingObject, message, ...values);
	}
	warn(mergingObject, message, ...values) {
		this.log("warn", mergingObject, message, ...values);
	}
	error(mergingObject, message, ...values) {
		this.log("error", mergingObject, message, ...values);
	}
	fatal(mergingObject, message, ...values) {
		this.log("fatal", mergingObject, message, ...values);
	}
	silent(mergingObject, message, ...values) {
		this.log("silent", mergingObject, message, ...values);
	}
	child(bindings, options) {
		if (!this.isEnabled) return this;
		return new Logger(this.config, this.pino.child(bindings, options));
	}
	bindings() {
		if (!this.isEnabled) return {};
		return this.pino.bindings();
	}
};
export { stdTimeFunctions as a, stdSerializers as i, destination as n, transport as o, multistream as r, Logger as t };
