// @adonisjs/core/build/commands/add.js
import "../chunk-iKc69rpz.js";
import "../main-kn40V-hF.js";
import "../debug-CGQmxzGt.js";
import "../main-Cxz0TyIw.js";
import "../config_provider-FIAUgvae.js";
import "../utils-rRkbAPnP.js";
import { n as prettyPrintError } from "../core-IpCOCkUL.js";
import { f as BaseCommand, l as flags, o as args } from "../main-MBAMnmJb.js";
import { t as __decorate } from "../decorate-DmrZA614.js";
var Add = class extends BaseCommand {
	static commandName = "add";
	static description = "Install and configure one or more packages";
	static options = { allowUnknownFlags: true };
	#resolveNpmPackageName(name) {
		if (name === "vinejs") return "@vinejs/vine";
		if (name === "edge") return "edge.js";
		return name;
	}
	async #configurePackage(packageName) {
		const flagValueArray = this.parsed.unknownFlags.filter((flag) => !!this.parsed.flags[flag]).map((flag) => [`--${flag}`, this.parsed.flags[flag].toString()]);
		const configureArgs = [
			packageName,
			this.force ? "--force" : void 0,
			this.verbose ? "--verbose" : void 0,
			...flagValueArray.flat()
		].filter(Boolean);
		return await this.kernel.exec("configure", configureArgs);
	}
	async run() {
		const packages = this.names.map((name) => ({
			name,
			npmName: this.#resolveNpmPackageName(name)
		}));
		const codemods = await this.createCodemods();
		codemods.verboseInstallOutput = !!this.verbose;
		if (!await codemods.installPackages(packages.map((pkg) => ({
			name: pkg.npmName,
			isDevDependency: !!this.dev
		})), this.packageManager)) return;
		const succeeded = [];
		const failed = [];
		for (const pkg of packages) {
			const { exitCode, error } = await this.#configurePackage(pkg.name);
			if (exitCode === 0) succeeded.push(pkg.name);
			else failed.push({
				name: pkg.name,
				error
			});
		}
		if (succeeded.length > 0) {
			const names = succeeded.map((name) => this.colors.green(name)).join(", ");
			this.logger.success(`Installed and configured ${names}`);
		}
		if (failed.length > 0) {
			this.exitCode = 1;
			for (const pkg of failed) {
				this.logger.error(`Unable to configure ${this.colors.green(pkg.name)}`);
				if (pkg.error) await prettyPrintError(pkg.error.cause || pkg.error);
			}
		}
	}
};
__decorate([args.spread({
	description: "Package names to install and configure",
	required: true
})], Add.prototype, "names", void 0);
__decorate([flags.boolean({ description: "Display logs in verbose mode" })], Add.prototype, "verbose", void 0);
__decorate([flags.string({ description: "Define the package manager you want to use" })], Add.prototype, "packageManager", void 0);
__decorate([flags.boolean({
	description: "Should we install the package as a dev dependency",
	alias: "D"
})], Add.prototype, "dev", void 0);
__decorate([flags.boolean({ description: "Forcefully overwrite existing files" })], Add.prototype, "force", void 0);
export { Add as default };
