// @adonisjs/core/build/commands/configure.js
import { t as stubsRoot } from "../main-kn40V-hF.js";
import { f as BaseCommand, l as flags, o as args } from "../main-MBAMnmJb.js";
import { t as __decorate } from "../decorate-DmrZA614.js";
import { RuntimeException } from "@poppinss/utils/exception";
var Configure = class extends BaseCommand {
	static commandName = "configure";
	static description = "Configure a package after it has been installed";
	static options = { allowUnknownFlags: true };
	get parsedFlags() {
		return this.parsed.flags;
	}
	get parsedArgs() {
		return this.parsed._;
	}
	async #getPackageSource(packageName) {
		try {
			return await this.app.import(packageName);
		} catch (error) {
			if (error.code && error.code === "ERR_MODULE_NOT_FOUND" || error.message.startsWith("Cannot find module")) return null;
			throw error;
		}
	}
	async #configureVineJS() {
		await (await this.createCodemods()).updateRcFile((rcFile) => {
			rcFile.addProvider("@adonisjs/core/providers/vinejs_provider");
		});
	}
	async #configureEdge() {
		await (await this.createCodemods()).updateRcFile((rcFile) => {
			rcFile.addProvider("@adonisjs/core/providers/edge_provider");
			rcFile.addMetaFile("resources/views/**/*.edge", false);
		});
	}
	async #configureHealthChecks() {
		const codemods = await this.createCodemods();
		await codemods.makeUsingStub(stubsRoot, "make/health/main.stub", {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity("health")
		});
		await codemods.makeUsingStub(stubsRoot, "make/health/controller.stub", {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity("health_checks")
		});
	}
	async createCodemods() {
		const codemods = await super.createCodemods();
		codemods.overwriteExisting = this.force === true;
		codemods.verboseInstallOutput = this.verbose === true;
		return codemods;
	}
	async run() {
		if (this.name === "vinejs") return this.#configureVineJS();
		if (this.name === "edge") return this.#configureEdge();
		if (this.name === "health_checks") return this.#configureHealthChecks();
		const packageExports = await this.#getPackageSource(this.name);
		if (!packageExports) {
			this.logger.error(`Cannot find module "${this.name}". Make sure to install it`);
			this.exitCode = 1;
			return;
		}
		if (!packageExports.configure) {
			this.logger.warning(`Cannot configure module "${this.name}". The module does not export the configure hook`);
			return;
		}
		if (packageExports.stubsRoot) this.stubsRoot = packageExports.stubsRoot;
		try {
			await packageExports.configure(this);
		} catch (error) {
			throw new RuntimeException(`Unable to configure package "${this.name}"`, { cause: error });
		}
	}
};
__decorate([args.string({ description: "Package name" })], Configure.prototype, "name", void 0);
__decorate([flags.boolean({
	description: "Display logs in verbose mode",
	alias: "v"
})], Configure.prototype, "verbose", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], Configure.prototype, "force", void 0);
export { Configure as default };
