// @adonisjs/core/build/commands/serve.js
import { t as importAssembler } from "../utils-rRkbAPnP.js";
import { f as BaseCommand, l as flags } from "../main-MBAMnmJb.js";
import { t as __decorate } from "../decorate-DmrZA614.js";
var Serve = class extends BaseCommand {
	static commandName = "serve";
	static description = "Start the development HTTP server along with the file watcher to perform restarts on file change";
	static help = [
		"Start the development server with file watcher using the following command.",
		"```",
		"{{ binaryName }} serve --watch",
		"```",
		"",
		"You can also start the server with HMR support using the following command.",
		"```",
		"{{ binaryName }} serve --hmr",
		"```",
		"",
		"The assets bundler dev server runs automatically after detecting vite config or webpack config files",
		"You may pass vite CLI args using the --assets-args command line flag.",
		"```",
		"{{ binaryName }} serve --assets-args=\"--debug --base=/public\"",
		"```"
	];
	static options = { staysAlive: true };
	#logMissingDevelopmentDependency(dependency) {
		this.logger.error([
			`Cannot find package "${dependency}"`,
			"",
			`The "${dependency}" package is a development dependency and therefore you should use the serve command during development only.`,
			"",
			"If you are running your application in production, then use \"node bin/server.js\" command to start the HTTP server"
		].join("\n"));
	}
	async run() {
		const assembler = await importAssembler(this.app);
		if (!assembler) {
			this.#logMissingDevelopmentDependency("@adonisjs/assembler");
			this.exitCode = 1;
			return;
		}
		if (this.watch && this.hmr) {
			this.logger.error("Cannot use --watch and --hmr flags together. Choose one of them");
			this.exitCode = 1;
			return;
		}
		this.devServer = new assembler.DevServer(this.app.appRoot, {
			hmr: this.hmr === true ? true : false,
			clearScreen: this.clear === false ? false : true,
			nodeArgs: this.parsed.nodeArgs,
			scriptArgs: [],
			metaFiles: this.app.rcFile.metaFiles,
			hooks: this.app.rcFile.hooks
		});
		this.devServer.ui.logger = this.logger;
		this.devServer.onClose((exitCode) => {
			this.exitCode = exitCode;
			this.terminate();
		});
		this.devServer.onError(() => {
			this.exitCode = 1;
			this.terminate();
		});
		if (this.watch) await this.devServer.startAndWatch({ poll: this.poll || false });
		else await this.devServer.start();
	}
};
__decorate([flags.boolean({ description: "Start the server with HMR support" })], Serve.prototype, "hmr", void 0);
__decorate([flags.boolean({
	description: "Watch filesystem and restart the HTTP server on file change",
	alias: "w"
})], Serve.prototype, "watch", void 0);
__decorate([flags.boolean({
	description: "Use polling to detect filesystem changes",
	alias: "p"
})], Serve.prototype, "poll", void 0);
__decorate([flags.boolean({
	description: "Clear the terminal for new logs after file change",
	showNegatedVariantInHelp: true,
	default: true
})], Serve.prototype, "clear", void 0);
export { Serve as default };
