// @adonisjs/application/build/application-BwkPOl3k.js
import { t as debug_default } from "./debug-DrQTxWHS.js";
import "node:module";
import { RuntimeException, createError } from "@poppinss/utils/exception";
import Hooks from "@poppinss/hooks";
import { fileURLToPath } from "node:url";
import { extname, join, relative } from "node:path";
import { Container } from "@adonisjs/fold";
import Macroable from "@poppinss/macroable";
import { detectAIAgent, importDefault, isRunningInAIAgent } from "@poppinss/utils";
import { inspect } from "node:util";
import string from "@poppinss/utils/string";
import StringBuilder from "@poppinss/utils/string_builder";
import { Config, ConfigLoader } from "@adonisjs/config";
import globParent from "glob-parent";
import diagnostics_channel from "node:diagnostics_channel";
var __defProp = Object.defineProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
var errors_exports = /* @__PURE__ */ __exportAll({
	E_INVALID_HOOKS_VALUE: () => E_INVALID_HOOKS_VALUE,
	E_INVALID_PRELOAD_FILE: () => E_INVALID_PRELOAD_FILE,
	E_INVALID_PRESETS_VALUE: () => E_INVALID_PRESETS_VALUE,
	E_INVALID_PRESET_FUNCTION: () => E_INVALID_PRESET_FUNCTION,
	E_INVALID_PROVIDER: () => E_INVALID_PROVIDER,
	E_MISSING_METAFILE_PATTERN: () => E_MISSING_METAFILE_PATTERN,
	E_MISSING_PRELOAD_FILE: () => E_MISSING_PRELOAD_FILE,
	E_MISSING_PROVIDER_FILE: () => E_MISSING_PROVIDER_FILE,
	E_MISSING_SUITE_FILES: () => E_MISSING_SUITE_FILES,
	E_MISSING_SUITE_NAME: () => E_MISSING_SUITE_NAME,
	E_PRESET_EXECUTION_ERROR: () => E_PRESET_EXECUTION_ERROR,
	E_UNKNOWN_ASSEMBLER_HOOK: () => E_UNKNOWN_ASSEMBLER_HOOK
});
const E_MISSING_METAFILE_PATTERN = createError("Invalid metafile entry \"%s\". Missing pattern property", "E_MISSING_METAFILE_PATTERN");
const E_MISSING_PRELOAD_FILE = createError("Invalid preload entry \"%s\". Missing file property", "E_MISSING_PRELOAD_FILE");
const E_INVALID_PRELOAD_FILE = createError("Invalid preload entry \"%s\". The file property must be a function", "E_INVALID_PRELOAD_FILE");
const E_MISSING_PROVIDER_FILE = createError("Invalid provider entry \"%s\". Missing file property", "E_MISSING_PROVIDER_FILE");
const E_INVALID_PROVIDER = createError("Invalid provider entry \"%s\". The file property must be a function", "E_INVALID_PROVIDER");
const E_MISSING_SUITE_NAME = createError("Invalid suite entry \"%s\". Missing name property", "E_MISSING_SUITE_NAME");
const E_MISSING_SUITE_FILES = createError("Invalid suite entry \"%s\". Missing files property", "E_MISSING_SUITE_FILES");
const E_UNKNOWN_ASSEMBLER_HOOK = createError("Assembler hook defined for unknown event \"%s\"", "E_UNKNOWN_ASSEMBLER_HOOK");
const E_INVALID_HOOKS_VALUE = createError("Expected hooks for event \"%s\" to be an array of dynamic imports. Instead received \"%s\"", "E_INVALID_HOOKS_VALUE");
const E_INVALID_PRESETS_VALUE = createError("Expected presets to be an array of functions. Instead received \"%s\"", "E_INVALID_PRESETS_VALUE");
const E_INVALID_PRESET_FUNCTION = createError("Expected preset at index %s to be a function. Instead received \"%s\"", "E_INVALID_PRESET_FUNCTION");
const E_PRESET_EXECUTION_ERROR = createError("Preset at index %s failed to execute: %s", "E_PRESET_EXECUTION_ERROR");
const generators = {
	singularControllerNames: [
		"home",
		"admin",
		"session",
		"application",
		"money",
		"signup",
		"login",
		"auth",
		"authentication",
		"adonis",
		"adonisjs",
		"dashboard",
		"api",
		"about",
		"contact",
		"blog",
		"library",
		"password",
		"password_link",
		"new_password",
		"profile",
		"account",
		"new_account",
		"avatar",
		"forum"
	],
	entityPathSegment(segmentName) {
		return new StringBuilder(segmentName).snakeCase().toString();
	},
	createEntity(entityName) {
		entityName = entityName.replace(new RegExp(`${extname(entityName)}$`), "");
		const parts = entityName.split("/");
		const [name] = parts.splice(-1);
		if (parts.length) return {
			path: parts.map((part) => this.entityPathSegment(part)).join("/"),
			name
		};
		return {
			path: "./",
			name
		};
	},
	importPath(...paths) {
		return string.toUnixSlash(join(...paths));
	},
	tableName(entityName) {
		return new StringBuilder(this.modelName(new StringBuilder(entityName).removeSuffix("table").toString())).plural().snakeCase().toString();
	},
	modelName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("model").singular().pascalCase().toString();
	},
	modelFileName(entityName) {
		return new StringBuilder(this.modelName(entityName)).snakeCase().ext(".ts").toString();
	},
	controllerName(entityName, singular = false) {
		const controller = new StringBuilder(entityName).removeExtension().removeSuffix("controller");
		if (this.singularControllerNames.includes(controller.toString().toLowerCase())) controller.singular();
		else if (singular) controller.singular();
		else controller.plural();
		return controller.pascalCase().suffix("Controller").toString();
	},
	controllerFileName(entityName, singular = false) {
		return new StringBuilder(this.controllerName(entityName, singular)).snakeCase().ext(".ts").toString();
	},
	eventName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("event").pascalCase().toString();
	},
	eventFileName(entityName) {
		return new StringBuilder(this.eventName(entityName)).snakeCase().ext(".ts").toString();
	},
	listenerName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("listener").pascalCase().toString();
	},
	listenerFileName(entityName) {
		return new StringBuilder(this.listenerName(entityName)).snakeCase().ext(".ts").toString();
	},
	middlewareName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("middleware").pascalCase().suffix("Middleware").toString();
	},
	middlewareFileName(entityName) {
		return new StringBuilder(this.middlewareName(entityName)).snakeCase().ext(".ts").toString();
	},
	providerName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("provider").singular().pascalCase().suffix("Provider").toString();
	},
	providerFileName(entityName) {
		return new StringBuilder(this.providerName(entityName)).snakeCase().ext(".ts").toString();
	},
	policyName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("policy").removeSuffix("model").singular().pascalCase().suffix("Policy").toString();
	},
	policyFileName(entityName) {
		return new StringBuilder(this.policyName(entityName)).snakeCase().ext(".ts").toString();
	},
	factoryName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("factory").removeSuffix("model").singular().pascalCase().suffix("Factory").toString();
	},
	factoryFileName(entityName) {
		return new StringBuilder(this.factoryName(entityName)).snakeCase().ext(".ts").toString();
	},
	serviceName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("service").removeSuffix("model").singular().pascalCase().suffix("Service").toString();
	},
	serviceFileName(entityName) {
		return new StringBuilder(this.serviceName(entityName)).snakeCase().ext(".ts").toString();
	},
	seederName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("seeder").removeSuffix("model").singular().pascalCase().suffix("Seeder").toString();
	},
	seederFileName(entityName) {
		return new StringBuilder(this.seederName(entityName)).snakeCase().ext(".ts").toString();
	},
	commandTerminalName(entityName) {
		const [namespace, ...rest] = new StringBuilder(this.commandName(entityName)).dashCase().toString().split("-");
		if (!rest.length) return namespace;
		return `${namespace}:${rest.join("-")}`;
	},
	commandName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("command").pascalCase().toString();
	},
	commandFileName(entityName) {
		return new StringBuilder(this.commandName(entityName)).snakeCase().ext(".ts").toString();
	},
	validatorName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("validator").singular().pascalCase().toString();
	},
	validatorActionName(entityName, action) {
		return new StringBuilder(this.validatorName(entityName)).prefix(`${action}_`).suffix("_validator").camelCase().toString();
	},
	validatorFileName(entityName) {
		return new StringBuilder(this.validatorName(entityName)).snakeCase().ext(".ts").toString();
	},
	exceptionName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("exception").pascalCase().suffix("Exception").toString();
	},
	exceptionFileName(entityName) {
		return new StringBuilder(this.exceptionName(entityName)).snakeCase().ext(".ts").toString();
	},
	mailerName(entityName, type = "notification") {
		return new StringBuilder(entityName).removeExtension().removeSuffix("notification").removeSuffix("provision").removeSuffix("mailer").pascalCase().suffix(string.pascalCase(type)).toString();
	},
	mailerFileName(entityName, type = "notification") {
		return new StringBuilder(this.mailerName(entityName, type)).snakeCase().ext(".ts").toString();
	},
	mailName(entityName, type = "notification") {
		return new StringBuilder(entityName).removeExtension().removeSuffix(type).removeSuffix("mailer", { similarWords: ["emailer"] }).removeSuffix("mail", { similarWords: ["email"] }).pascalCase().suffix(string.pascalCase(type)).toString();
	},
	mailFileName(entityName, type = "notification") {
		return new StringBuilder(this.mailName(entityName, type)).snakeCase().ext(".ts").toString();
	},
	testGroupName(entity) {
		return new StringBuilder(`${entity.path}/${entity.name}`).removeExtension().removeSuffix(".spec").sentenceCase().toString();
	},
	testFileName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix(".spec").snakeCase().ext(".spec.ts").toString();
	},
	viewFileName(entityName) {
		return new StringBuilder(entityName).removeExtension().snakeCase().ext(".edge").toString();
	},
	transformerName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("transformer").removeSuffix("model").singular().pascalCase().suffix("Transformer").toString();
	},
	transformerFileName(entityName) {
		return new StringBuilder(this.transformerName(entityName)).snakeCase().ext(".ts").toString();
	},
	inertiaPageName(entityName) {
		return new StringBuilder(entityName).removeExtension().removeSuffix("page").pascalCase().toString();
	},
	inertiaPageFileName(entityName, extension) {
		return new StringBuilder(this.inertiaPageName(entityName)).snakeCase().ext(extension).toString();
	}
};
var ConfigManager = class {
	#appRoot;
	#configValues;
	config;
	constructor(appRoot) {
		this.#appRoot = appRoot;
	}
	useConfig(values) {
		this.#configValues = values;
		return this;
	}
	async process(configDirectory) {
		if (this.#configValues) this.config = new Config(this.#configValues);
		else {
			const loader = new ConfigLoader(new URL(configDirectory, this.#appRoot));
			debug_default("loading config from directory \"%s\"", configDirectory);
			this.config = new Config(await loader.load());
		}
		this.#configValues = void 0;
	}
};
const directories = {
	config: "config",
	commands: "commands",
	contracts: "contracts",
	public: "public",
	providers: "providers",
	languageFiles: "resources/lang",
	migrations: "database/migrations",
	seeders: "database/seeders",
	factories: "database/factories",
	views: "resources/views",
	start: "start",
	tmp: "tmp",
	tests: "tests",
	httpControllers: "app/controllers",
	models: "app/models",
	services: "app/services",
	exceptions: "app/exceptions",
	mailers: "app/mailers",
	mails: "app/mails",
	middleware: "app/middleware",
	policies: "app/policies",
	validators: "app/validators",
	events: "app/events",
	listeners: "app/listeners",
	transformers: "app/transformers",
	stubs: "stubs",
	generatedClient: ".adonisjs/client",
	generatedServer: ".adonisjs/server"
};
const KNOWN_ASSEMBLER_HOOKS = [
	"init",
	"routesCommitted",
	"routesScanning",
	"routesScanned",
	"routesCommitted",
	"buildStarting",
	"buildFinished",
	"devServerStarting",
	"devServerStarted",
	"testsStarting",
	"testsFinished",
	"fileAdded",
	"fileChanged",
	"fileRemoved"
];
var RcFileParser = class {
	#defaults = {
		typescript: true,
		preloads: [],
		metaFiles: [],
		presets: [],
		commandsAliases: {},
		commands: [],
		providers: [],
		directories,
		tests: {
			suites: [],
			timeout: 2e3,
			forceExit: true
		},
		hooks: {},
		experimental: {}
	};
	#rcFile;
	#raw;
	constructor(rcFile) {
		this.#rcFile = Object.assign(this.#defaults, rcFile);
		this.#raw = rcFile;
	}
	#knownEnvironments() {
		return [
			"web",
			"console",
			"test",
			"repl"
		];
	}
	#getHooks() {
		const hooks = this.#rcFile.hooks;
		if (!hooks) return;
		return Object.keys(hooks).reduce((result, eventName) => {
			if (!KNOWN_ASSEMBLER_HOOKS.includes(eventName)) throw new E_UNKNOWN_ASSEMBLER_HOOK([eventName]);
			const eventHooks = hooks[eventName];
			if (!Array.isArray(eventHooks)) throw new E_INVALID_HOOKS_VALUE([eventName, inspect(eventHooks)]);
			result[eventName] = eventHooks;
			return result;
		}, {});
	}
	#getPreloads() {
		return this.#rcFile.preloads.map((preload) => {
			const normalizedPreload = typeof preload === "function" ? {
				file: preload,
				environment: this.#knownEnvironments()
			} : preload;
			if (!normalizedPreload.file) throw new E_MISSING_PRELOAD_FILE([inspect(preload)]);
			if (typeof normalizedPreload.file !== "function") throw new E_INVALID_PRELOAD_FILE([inspect(preload)]);
			return {
				file: normalizedPreload.file,
				environment: normalizedPreload.environment ?? this.#knownEnvironments()
			};
		});
	}
	#getProviders() {
		return this.#rcFile.providers.map((provider) => {
			const normalizedProvider = typeof provider === "function" ? {
				file: provider,
				environment: this.#knownEnvironments()
			} : provider;
			if (!normalizedProvider.file) throw new E_MISSING_PROVIDER_FILE([inspect(provider)]);
			if (typeof normalizedProvider.file !== "function") throw new E_INVALID_PROVIDER([inspect(provider)]);
			return {
				file: normalizedProvider.file,
				environment: normalizedProvider.environment ?? this.#knownEnvironments()
			};
		});
	}
	#getMetaFiles() {
		return this.#rcFile.metaFiles.map((pattern) => {
			const normalizeMetaFile = typeof pattern === "string" ? {
				pattern,
				reloadServer: true
			} : pattern;
			if (!normalizeMetaFile.pattern) throw new E_MISSING_METAFILE_PATTERN([inspect(pattern)]);
			return {
				pattern: normalizeMetaFile.pattern,
				reloadServer: normalizeMetaFile.reloadServer ?? true
			};
		});
	}
	#getSuites() {
		return (this.#rcFile.tests.suites || []).map((suite) => {
			if (!suite.name) throw new E_MISSING_SUITE_NAME([inspect(suite)]);
			if (!suite.files) throw new E_MISSING_SUITE_FILES([inspect(suite)]);
			const files = Array.isArray(suite.files) ? [...suite.files] : [suite.files];
			return {
				name: suite.name,
				files,
				directories: files.map((file) => globParent(file)),
				timeout: suite.timeout
			};
		});
	}
	#applyPresets() {
		if (this.#rcFile.presets.length === 0) return;
		if (!Array.isArray(this.#rcFile.presets)) throw new E_INVALID_PRESETS_VALUE([inspect(this.#rcFile.presets)]);
		this.#rcFile.presets.forEach((preset, index) => {
			if (typeof preset !== "function") throw new E_INVALID_PRESET_FUNCTION([index, inspect(preset)]);
			try {
				preset({ rcFile: this.#rcFile });
			} catch (error) {
				throw new E_PRESET_EXECUTION_ERROR([index, error.message]);
			}
		});
	}
	parse() {
		this.#applyPresets();
		return {
			typescript: this.#rcFile.typescript,
			preloads: this.#getPreloads(),
			metaFiles: this.#getMetaFiles(),
			commands: [...this.#rcFile.commands],
			directories: {
				...directories,
				...this.#rcFile.directories
			},
			commandsAliases: { ...this.#rcFile.commandsAliases },
			providers: this.#getProviders(),
			tests: {
				suites: this.#getSuites(),
				timeout: this.#rcFile.tests.timeout ?? 2e3,
				forceExit: this.#rcFile.tests.forceExit ?? true
			},
			hooks: this.#getHooks(),
			experimental: this.#rcFile.experimental,
			raw: this.#raw
		};
	}
};
var RcFileManager = class {
	#appRoot;
	#rcContents;
	rcFile;
	constructor(appRoot) {
		this.#appRoot = appRoot;
	}
	rcContents(value) {
		this.#rcContents = value;
		return this;
	}
	async process() {
		if (!this.#rcContents) {
			const rcTSFile = new URL("adonisrc.js", this.#appRoot);
			try {
				this.#rcContents = (await import(rcTSFile.href)).default;
				debug_default("adonisrc.ts file contents: %O", this.#rcContents);
			} catch (error) {
				if (!/Cannot find module/.test(error.message)) throw error;
			}
		}
		this.rcFile = new RcFileParser(this.#rcContents).parse();
		this.#rcContents = void 0;
	}
};
const TEST_ENVS = ["test", "testing"];
const PROD_ENVS = ["prod", "production"];
const DEV_ENVS = [
	"dev",
	"develop",
	"development"
];
var NodeEnvManager = class {
	nodeEnvironment = "unknown";
	#normalizeNodeEnv(env) {
		if (!env || typeof env !== "string") return "unknown";
		env = env.toLowerCase();
		if (DEV_ENVS.includes(env)) return "development";
		if (PROD_ENVS.includes(env)) return "production";
		if (TEST_ENVS.includes(env)) return "test";
		return env;
	}
	process() {
		this.nodeEnvironment = this.#normalizeNodeEnv(process.env.NODE_ENV);
	}
};
var tracing_channels_exports = /* @__PURE__ */ __exportAll({
	preloadImport: () => preloadImport,
	providerBoot: () => providerBoot,
	providerReady: () => providerReady,
	providerRegister: () => providerRegister,
	providerShutdown: () => providerShutdown,
	providerStart: () => providerStart
});
const providerRegister = diagnostics_channel.tracingChannel("adonisjs.provider.register");
const providerBoot = diagnostics_channel.tracingChannel("adonisjs.provider.boot");
const providerStart = diagnostics_channel.tracingChannel("adonisjs.provider.start");
const providerReady = diagnostics_channel.tracingChannel("adonisjs.provider.ready");
const providerShutdown = diagnostics_channel.tracingChannel("adonisjs.provider.shutdown");
const preloadImport = diagnostics_channel.tracingChannel("adonisjs.preload.import");
var PreloadsManager = class {
	#list = [];
	#options;
	constructor(options) {
		this.#options = options;
	}
	#filterByEnvironment(provider) {
		if (this.#options.environment === "unknown") return false;
		return provider.environment.includes(this.#options.environment);
	}
	use(list) {
		this.#list = list;
		return this;
	}
	setEnvironment(environment) {
		debug_default("switching environment for preloads { from:\"%s\", to: \"%s\" }", this.#options.environment, environment);
		this.#options.environment = environment;
		return this;
	}
	async import() {
		const preloads = this.#list.filter((preload) => this.#filterByEnvironment(preload));
		debug_default("preloading modules %O", preloads);
		await Promise.all(preloads.map((preload) => preloadImport.tracePromise(preload.file, preloadImport.hasSubscribers ? { file: preload.file } : void 0, preload)));
		this.#list = [];
	}
};
var ProvidersManager = class {
	#providers = [];
	#providersWithShutdownListeners = [];
	#list = [];
	#options;
	constructor(options) {
		this.#options = options;
	}
	#filterByEnvironment(provider) {
		if (this.#options.environment === "unknown") return false;
		return provider.environment.includes(this.#options.environment);
	}
	#isAClass(providerClass) {
		return typeof providerClass === "function" && providerClass.toString().startsWith("class ");
	}
	async #resolveProvider(provider) {
		const providerExports = await provider.file();
		if (Object.keys(providerExports).length === 0) return null;
		if (!providerExports.default) throw new RuntimeException(`Invalid exports from "${provider.file.toString()}" provider. It must have a default export`);
		if (!this.#isAClass(providerExports.default)) throw new RuntimeException(`Default export from module "${provider.file.toString()}" is not a class`);
		return providerExports.default;
	}
	#resolve() {
		const providers = this.#list.filter((provider) => this.#filterByEnvironment(provider));
		debug_default("loading providers %O", providers);
		return Promise.all(providers.map((provider) => this.#resolveProvider(provider)));
	}
	use(list) {
		this.#list = list;
		return this;
	}
	setEnvironment(environment) {
		debug_default("switching environment for providers { from:\"%s\", to: \"%s\" }", this.#options.environment, environment);
		this.#options.environment = environment;
		return this;
	}
	async register() {
		const providers = await this.#resolve();
		this.#list = [];
		providers.forEach((provider) => {
			if (provider) {
				debug_default("registering \"%s\"", provider.name);
				const providerInstance = new provider(...this.#options.providersState);
				this.#providers.push(providerInstance);
				if (providerInstance.shutdown) this.#providersWithShutdownListeners.push(providerInstance);
				if (providerInstance.register) providerRegister.traceSync(providerInstance.register, providerRegister.hasSubscribers ? { provider: providerInstance } : void 0, providerInstance);
			}
		});
	}
	async boot() {
		for (let provider of this.#providers) if (provider.boot) {
			debug_default("booting \"%s\"", provider.constructor.name);
			await providerBoot.tracePromise(provider.boot, providerBoot.hasSubscribers ? { provider } : void 0, provider);
		}
	}
	async start() {
		for (let provider of this.#providers) if (provider.start) {
			debug_default("invoking \"%s\" start method", provider.constructor.name);
			await providerStart.tracePromise(provider.start, providerStart.hasSubscribers ? { provider } : void 0, provider);
		}
	}
	async ready() {
		for (let provider of this.#providers) if (provider.ready) {
			debug_default("invoking \"%s\" ready method", provider.constructor.name);
			await providerReady.tracePromise(provider.ready, providerReady.hasSubscribers ? { provider } : void 0, provider);
		}
		this.#providers = [];
	}
	async shutdown(inReverseOrder) {
		const providersWithShutdownListeners = inReverseOrder ? Array.from(this.#providersWithShutdownListeners).reverse() : Array.from(this.#providersWithShutdownListeners);
		this.#providersWithShutdownListeners = [];
		for (let provider of providersWithShutdownListeners) if (provider.shutdown) {
			debug_default("shutting down \"%s\"", provider.constructor.name);
			await providerShutdown.tracePromise(provider.shutdown, providerShutdown.hasSubscribers ? { provider } : void 0, provider);
		}
	}
};
var FeatureFlags = class {
	#flags;
	#flagsFactory;
	constructor(flags) {
		if (typeof flags === "function") this.#flagsFactory = flags;
		else this.#flags = flags;
	}
	enabled(feature) {
		return (this.#flags ?? this.#flagsFactory())[feature] === true;
	}
	disabled(feature) {
		return (this.#flags ?? this.#flagsFactory())[feature] === false;
	}
	has(feature) {
		return feature in (this.#flags ?? this.#flagsFactory());
	}
	when(feature, enabledCallback, disabledCallback) {
		if (this.enabled(feature)) return enabledCallback();
		return disabledCallback ? disabledCallback() : void 0;
	}
};
var Application = class extends Macroable {
	#importer;
	#terminating = false;
	#surroundedEnvironment = { pm2: false };
	#appRoot;
	#environment;
	#state = "created";
	#configManager;
	#rcFileManager;
	#nodeEnvManager;
	#preloadsManager;
	#providersManager;
	#hooks = new Hooks();
	info = /* @__PURE__ */ new Map();
	get appName() {
		return this.info.get("appName") || "adonisjs_app";
	}
	get version() {
		return this.info.get("version") || null;
	}
	get adonisVersion() {
		return this.info.get("adonisVersion") || null;
	}
	get appRoot() {
		return this.#appRoot;
	}
	get isBooted() {
		return this.#state !== "created" && this.#state !== "initiated";
	}
	get isReady() {
		return this.#state === "ready";
	}
	get isTerminated() {
		return this.#state === "terminated";
	}
	get isTerminating() {
		return this.#terminating && this.#state !== "terminated";
	}
	get config() {
		return this.#configManager.config;
	}
	get rcFile() {
		return this.#rcFileManager.rcFile;
	}
	get nodeEnvironment() {
		return this.#nodeEnvManager.nodeEnvironment;
	}
	get inProduction() {
		return this.nodeEnvironment === "production";
	}
	get inDev() {
		return this.nodeEnvironment === "development";
	}
	get inTest() {
		return this.nodeEnvironment === "test";
	}
	get managedByPm2() {
		return this.#surroundedEnvironment.pm2;
	}
	get generators() {
		return generators;
	}
	stubs = { create: async () => {
		const { StubsManager } = await import("./manager-BXPolBfe.js");
		return new StubsManager(this, this.makePath(this.rcFile.directories.stubs));
	} };
	experimentalFlags = new FeatureFlags(() => this.#rcFileManager.rcFile.experimental);
	usingVineJS = false;
	usingEdgeJS = false;
	get detectedAIAgent() {
		return detectAIAgent();
	}
	get runningInAIAgent() {
		return isRunningInAIAgent();
	}
	constructor(appRoot, options) {
		super();
		this.#appRoot = appRoot;
		this.#importer = options.importer;
		this.#environment = options.environment;
		this.#nodeEnvManager = new NodeEnvManager();
		this.#configManager = new ConfigManager(this.appRoot);
		this.#rcFileManager = new RcFileManager(this.appRoot);
		this.#providersManager = new ProvidersManager({
			environment: this.#environment,
			providersState: [this]
		});
		this.#preloadsManager = new PreloadsManager({ environment: this.#environment });
		this.#surroundedEnvironment.pm2 = !!process.env.pm2_id;
		if (debug_default.enabled) debug_default("app environment :%O", {
			pm2: this.#surroundedEnvironment.pm2,
			environment: this.#environment,
			nodeEnv: this.#nodeEnvManager.nodeEnvironment
		});
	}
	#instantiateContainer() {
		this.container = new Container();
	}
	getEnvironment() {
		return this.#environment;
	}
	setEnvironment(environment) {
		if (this.#state !== "created" && this.#state !== "initiated") throw new RuntimeException("Cannot switch environment once the app has been booted");
		debug_default("switching environment { from:\"%s\", to: \"%s\" }", this.#environment, environment);
		this.#environment = environment;
		this.#preloadsManager.setEnvironment(environment);
		this.#providersManager.setEnvironment(environment);
		return this;
	}
	getState() {
		return this.#state;
	}
	rcContents(value) {
		this.#rcFileManager.rcContents(value);
		return this;
	}
	useConfig(values) {
		this.#configManager.useConfig(values);
		return this;
	}
	notify(message, sendHandle, options, callback) {
		if (process.send) process.send(message, sendHandle, options, callback);
	}
	listen(signal, callback) {
		process.on(signal, callback);
		return this;
	}
	listenOnce(signal, callback) {
		process.once(signal, callback);
		return this;
	}
	listenIf(conditional, signal, callback) {
		if (conditional) process.on(signal, callback);
		return this;
	}
	listenOnceIf(conditional, signal, callback) {
		if (conditional) process.once(signal, callback);
		return this;
	}
	initiating(handler) {
		this.#hooks.add("initiating", handler);
		return this;
	}
	async init() {
		if (this.#state !== "created") {
			debug_default("cannot initiate app from state \"%s\"", this.#state);
			return;
		}
		debug_default("initiating app");
		this.#instantiateContainer();
		await this.#hooks.runner("initiating").run(this);
		await this.#rcFileManager.process();
		this.#hooks.clear("initiating");
		this.#state = "initiated";
	}
	booting(handler) {
		this.#hooks.add("booting", handler);
		return this;
	}
	async boot() {
		if (this.#state !== "initiated") {
			debug_default("cannot boot app from state \"%s\"", this.#state);
			return;
		}
		debug_default("booting app");
		await this.#hooks.runner("booting").run(this);
		this.#hooks.clear("booting");
		this.#nodeEnvManager.process();
		await this.#configManager.process(this.rcFile.directories.config);
		this.#providersManager.use(this.rcFile.providers);
		await this.#providersManager.register();
		await this.#providersManager.boot();
		await this.#hooks.runner("booted").run(this);
		this.#hooks.clear("booted");
		this.#state = "booted";
	}
	async booted(handler) {
		if (this.isBooted) await handler(this);
		else this.#hooks.add("booted", handler);
	}
	starting(handler) {
		this.#hooks.add("starting", handler);
		return this;
	}
	async start(callback) {
		if (this.#state !== "booted") {
			debug_default("cannot start app from state \"%s\"", this.#state);
			return;
		}
		debug_default("starting app");
		await this.#providersManager.start();
		await this.#hooks.runner("starting").run(this);
		this.#hooks.clear("starting");
		await this.#preloadsManager.use(this.rcFile.preloads).import();
		await callback(this);
		await this.#providersManager.ready();
		await this.#hooks.runner("ready").run(this);
		this.#hooks.clear("ready");
		this.#state = "ready";
		debug_default("application ready");
		this.notify("ready");
	}
	async ready(handler) {
		if (this.isReady) await handler(this);
		else this.#hooks.add("ready", handler);
	}
	terminating(handler) {
		this.#hooks.add("terminating", handler);
		return this;
	}
	async terminate() {
		if (this.#state === "created" || this.#state === "terminated") {
			debug_default("cannot terminate app from state \"%s\"", this.#state);
			return;
		}
		if (this.#terminating) {
			debug_default("app is already being terminated");
			return;
		}
		debug_default("terminating app");
		this.#terminating = true;
		await this.#hooks.runner("terminating").runReverse(this);
		await this.#providersManager.shutdown(true);
		this.#hooks.clear("terminating");
		this.#state = "terminated";
	}
	relativePath(absolutePath) {
		return relative(fileURLToPath(this.appRoot), absolutePath);
	}
	makeURL(...paths) {
		return new URL(join(...paths), this.#appRoot);
	}
	makePath(...paths) {
		return fileURLToPath(this.makeURL(...paths));
	}
	configPath(...paths) {
		return this.makePath(this.rcFile.directories.config, ...paths);
	}
	publicPath(...paths) {
		return this.makePath(this.rcFile.directories.public, ...paths);
	}
	providersPath(...paths) {
		return this.makePath(this.rcFile.directories.providers, ...paths);
	}
	factoriesPath(...paths) {
		return this.makePath(this.rcFile.directories.factories, ...paths);
	}
	migrationsPath(...paths) {
		return this.makePath(this.rcFile.directories.migrations, ...paths);
	}
	seedersPath(...paths) {
		return this.makePath(this.rcFile.directories.seeders, ...paths);
	}
	languageFilesPath(...paths) {
		return this.makePath(this.rcFile.directories.languageFiles, ...paths);
	}
	viewsPath(...paths) {
		return this.makePath(this.rcFile.directories.views, ...paths);
	}
	startPath(...paths) {
		return this.makePath(this.rcFile.directories.start, ...paths);
	}
	tmpPath(...paths) {
		return this.makePath(this.rcFile.directories.tmp, ...paths);
	}
	contractsPath(...paths) {
		return this.makePath(this.rcFile.directories.contracts, ...paths);
	}
	httpControllersPath(...paths) {
		return this.makePath(this.rcFile.directories.httpControllers, ...paths);
	}
	modelsPath(...paths) {
		return this.makePath(this.rcFile.directories.models, ...paths);
	}
	servicesPath(...paths) {
		return this.makePath(this.rcFile.directories.services, ...paths);
	}
	exceptionsPath(...paths) {
		return this.makePath(this.rcFile.directories.exceptions, ...paths);
	}
	mailersPath(...paths) {
		return this.makePath(this.rcFile.directories.mailers, ...paths);
	}
	mailsPath(...paths) {
		return this.makePath(this.rcFile.directories.mails, ...paths);
	}
	middlewarePath(...paths) {
		return this.makePath(this.rcFile.directories.middleware, ...paths);
	}
	policiesPath(...paths) {
		return this.makePath(this.rcFile.directories.policies, ...paths);
	}
	validatorsPath(...paths) {
		return this.makePath(this.rcFile.directories.validators, ...paths);
	}
	commandsPath(...paths) {
		return this.makePath(this.rcFile.directories.commands, ...paths);
	}
	eventsPath(...paths) {
		return this.makePath(this.rcFile.directories.events, ...paths);
	}
	listenersPath(...paths) {
		return this.makePath(this.rcFile.directories.listeners, ...paths);
	}
	transformersPath(...paths) {
		return this.makePath(this.rcFile.directories.transformers, ...paths);
	}
	generatedClientPath(...paths) {
		return this.makePath(this.rcFile.directories.generatedClient, ...paths);
	}
	generatedServerPath(...paths) {
		return this.makePath(this.rcFile.directories.generatedServer, ...paths);
	}
	import(moduleIdentifier) {
		if (!this.#importer) throw new RuntimeException(`Cannot import "${moduleIdentifier}". Register a module importer with the application first.`);
		return this.#importer(moduleIdentifier);
	}
	importDefault(moduleIdentifier) {
		if (!this.#importer) throw new RuntimeException(`Cannot import "${moduleIdentifier}". Register a module importer with the application first.`);
		return importDefault(() => this.#importer(moduleIdentifier));
	}
	toJSON() {
		return {
			isReady: this.isReady,
			isTerminating: this.isTerminating,
			environment: this.#environment,
			nodeEnvironment: this.nodeEnvironment,
			appName: this.appName,
			version: this.version ? this.version.toString() : null,
			adonisVersion: this.adonisVersion ? this.adonisVersion.toString() : null
		};
	}
};
export { generators as a, RcFileParser as i, FeatureFlags as n, errors_exports as o, tracing_channels_exports as r, Application as t };
