// @adonisjs/application/build/index.js
import { a as generators, i as RcFileParser, n as FeatureFlags, o as errors_exports, r as tracing_channels_exports, t as Application } from "./application-BwkPOl3k.js";
import "./debug-DrQTxWHS.js";
const hooks = {
	init(callback) {
		return callback;
	},
	routesCommitted(callback) {
		return callback;
	},
	routesScanning(callback) {
		return callback;
	},
	routesScanned(callback) {
		return callback;
	},
	fileChanged(callback) {
		return callback;
	},
	fileAdded(callback) {
		return callback;
	},
	fileRemoved(callback) {
		return callback;
	},
	devServerStarting(callback) {
		return callback;
	},
	devServerStarted(callback) {
		return callback;
	},
	buildStarting(callback) {
		return callback;
	},
	buildFinished(callback) {
		return callback;
	},
	testsStarting(callback) {
		return callback;
	},
	testsFinished(callback) {
		return callback;
	}
};
function defineConfig(config) {
	return config;
}
export { Application, FeatureFlags, RcFileParser, defineConfig, errors_exports as errors, generators, hooks, tracing_channels_exports as tracingChannels };
