// @adonisjs/http-server/build/src/client/url_builder.js
import { n as findRoute, t as createURL } from "../../helpers-C_2HouOe.js";
import "../../types-AUwURgIL.js";
function createUrlBuilder(routesLoader, searchParamsStringifier) {
	let domainsList;
	let domainsRoutes;
	function createUrlForRoute(identifier, params, options, method) {
		if (!domainsRoutes) {
			domainsRoutes = typeof routesLoader === "function" ? routesLoader() : routesLoader;
			if (!domainsRoutes || typeof domainsRoutes !== "object") throw new Error(`Cannot construct routes. Expected the value to be an object, instead received ${typeof domainsRoutes}`);
			if (!domainsList) domainsList = Object.keys(domainsRoutes).filter((domain) => domain !== "root");
		}
		const domain = domainsList.find((name) => identifier.startsWith(`${name}@`));
		const routeIdentifier = domain ? identifier.replace(new RegExp(`^${domain}@`), "") : identifier;
		const route = findRoute(domainsRoutes, routeIdentifier, domain, method, true);
		if (!route) {
			if (method) throw new Error(`Cannot lookup route "${routeIdentifier}" for method "${method}"`);
			throw new Error(`Cannot lookup route "${routeIdentifier}"`);
		}
		return createURL(route.name ?? route.pattern, route.tokens, searchParamsStringifier, params, options);
	}
	const urlFor = function route(...[identifier, params, options]) {
		return createUrlForRoute(identifier, params, options);
	};
	urlFor.get = function urlForMethodGet(...[identifier, params, options]) {
		const method = "GET";
		const url = createUrlForRoute(identifier, params, options, method);
		return {
			url,
			method,
			toString() {
				return url;
			},
			form: {
				action: url,
				method
			}
		};
	};
	urlFor.post = function urlForMethodPost(...[identifier, params, options]) {
		const method = "POST";
		const url = createUrlForRoute(identifier, params, options, method);
		return {
			url,
			method,
			toString() {
				return url;
			},
			form: {
				action: url,
				method
			}
		};
	};
	urlFor.put = function urlForMethodPut(...[identifier, params, options]) {
		const method = "PUT";
		const url = createUrlForRoute(identifier, params, options, method);
		return {
			url,
			method,
			toString() {
				return url;
			},
			form: {
				action: url,
				method
			}
		};
	};
	urlFor.patch = function urlForMethodPatch(...[identifier, params, options]) {
		const method = "PATCH";
		const url = createUrlForRoute(identifier, params, options, method);
		return {
			url,
			method,
			toString() {
				return url;
			},
			form: {
				action: url,
				method
			}
		};
	};
	urlFor.delete = function urlForMethodDelete(...[identifier, params, options]) {
		const method = "DELETE";
		const url = createUrlForRoute(identifier, params, options, method);
		return {
			url,
			method,
			toString() {
				return url;
			},
			form: {
				action: url,
				method
			}
		};
	};
	urlFor.method = function urlForCustomMethod(method, ...[identifier, params, options]) {
		const url = createUrlForRoute(identifier, params, options, method);
		return {
			url,
			method,
			toString() {
				return url;
			},
			form: {
				action: url,
				method
			}
		};
	};
	return urlFor;
}
export { createURL, createUrlBuilder, findRoute };
