// @adonisjs/bodyparser/build/factories/main.js
import { t as MultipartFile } from "../file-CexcC_98.js";
import { t as defineConfig } from "../define_config-CKeeX9du.js";
import "../multipart-y_9_4MLm.js";
import { BodyParserMiddleware } from "../src/bodyparser_middleware.js";
import lodash from "@poppinss/utils/lodash";
var MultipartFileFactory = class {
	#parameters = {};
	merge(params) {
		this.#parameters = Object.assign(this.#parameters, params);
		return this;
	}
	create(validationOptions) {
		const file = new MultipartFile({
			fieldName: this.#parameters.fieldName || "file",
			clientName: this.#parameters.clientName || this.#parameters.extname ? `file.${this.#parameters.extname}` : "file",
			headers: this.#parameters.headers || {}
		}, validationOptions || {});
		file.size = this.#parameters.size || 0;
		file.extname = this.#parameters.extname;
		file.type = this.#parameters.type;
		file.subtype = this.#parameters.subtype;
		file.state = "consumed";
		file.validate();
		return file;
	}
};
var BodyParserMiddlewareFactory = class {
	#config = defineConfig({});
	#featureFlags;
	#getConfig() {
		return this.#config;
	}
	merge(config) {
		this.#config = lodash.merge(this.#config, config);
		return this;
	}
	withFeatureFlags(featureFlags) {
		this.#featureFlags = featureFlags;
		return this;
	}
	create() {
		return new BodyParserMiddleware(this.#getConfig(), this.#featureFlags);
	}
};
export { BodyParserMiddlewareFactory, MultipartFileFactory };
