// @adonisjs/core/build/commands/env/add.js
import stringHelpers from "../../src/helpers/string.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
const ALLOWED_TYPES = [
	"string",
	"boolean",
	"number",
	"enum"
];
var EnvAdd = class extends BaseCommand {
	static commandName = "env:add";
	static description = "Add a new environment variable";
	static options = { allowUnknownFlags: true };
	#isTypeFlagValid() {
		return ALLOWED_TYPES.includes(this.type);
	}
	async run() {
		if (!this.name) this.name = await this.prompt.ask("Enter the variable name", {
			validate: (value) => !!value,
			format: (value) => stringHelpers.snakeCase(value).toUpperCase()
		});
		if (!this.value) this.value = await this.prompt.ask("Enter the variable value");
		if (!this.type) this.type = await this.prompt.choice("Select the variable type", ALLOWED_TYPES);
		if (this.type === "enum" && !this.enumValues) this.enumValues = await this.prompt.ask("Enter the enum values separated by a comma", { result: (value) => value.split(",").map((one) => one.trim()) });
		if (!this.#isTypeFlagValid()) {
			this.logger.error(`Invalid type "${this.type}". Must be one of ${ALLOWED_TYPES.join(", ")}`);
			return;
		}
		const codemods = await this.createCodemods();
		const transformedName = stringHelpers.snakeCase(this.name).toUpperCase();
		await codemods.defineEnvVariables({ [transformedName]: this.value }, { omitFromExample: [transformedName] });
		const validation = {
			string: "Env.schema.string()",
			number: "Env.schema.number()",
			boolean: "Env.schema.boolean()",
			enum: `Env.schema.enum(['${this.enumValues.join("','")}'] as const)`
		}[this.type];
		await codemods.defineEnvValidations({ variables: { [transformedName]: validation } });
		this.logger.success("Environment variable added successfully");
	}
};
__decorate([args.string({
	description: "Variable name. Will be converted to screaming snake case",
	required: false
})], EnvAdd.prototype, "name", void 0);
__decorate([args.string({
	description: "Variable value",
	required: false
})], EnvAdd.prototype, "value", void 0);
__decorate([flags.string({ description: "Type of the variable" })], EnvAdd.prototype, "type", void 0);
__decorate([flags.array({
	description: "Allowed values for the enum type in a comma-separated list",
	default: [""],
	required: false
})], EnvAdd.prototype, "enumValues", void 0);
export { EnvAdd as default };
