// @adonisjs/core/build/commands/make/controller.js
import { t as stubsRoot } from "../../main-kn40V-hF.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
import string from "@poppinss/utils/string";
var MakeController = class extends BaseCommand {
	static commandName = "make:controller";
	static description = "Create a new HTTP controller class";
	static options = { allowUnknownFlags: true };
	stubPath = "make/controller/main.stub";
	async prepare() {
		if (this.actions) this.stubPath = "make/controller/actions.stub";
		if (this.resource) if (this.actions) this.logger.warning("Cannot use --resource flag with actions. Ignoring --resource");
		else this.stubPath = "make/controller/resource.stub";
		if (this.api) if (this.actions) this.logger.warning("Cannot use --api flag with actions. Ignoring --api");
		else this.stubPath = "make/controller/api.stub";
		if (this.resource && this.api && !this.actions) this.logger.warning("--api and --resource flags cannot be used together. Ignoring --resource");
	}
	async run() {
		await (await this.createCodemods()).makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			actions: this.actions?.map((action) => string.camelCase(action)),
			entity: this.app.generators.createEntity(this.name),
			singular: this.singular
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "The name of the controller" })], MakeController.prototype, "name", void 0);
__decorate([args.spread({
	description: "Create controller with custom method names",
	required: false
})], MakeController.prototype, "actions", void 0);
__decorate([flags.boolean({
	description: "Generate controller in singular form",
	alias: "s"
})], MakeController.prototype, "singular", void 0);
__decorate([flags.boolean({
	description: "Generate resourceful controller with methods to perform CRUD actions on a resource",
	alias: "r"
})], MakeController.prototype, "resource", void 0);
__decorate([flags.boolean({
	description: "Generate resourceful controller without the \"edit\" and the \"create\" methods",
	alias: "a"
})], MakeController.prototype, "api", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeController.prototype, "contentsFrom", void 0);
export { MakeController as default };
