// @adonisjs/core/build/commands/make/listener.js
import { t as stubsRoot } from "../../main-kn40V-hF.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
var MakeListener = class extends BaseCommand {
	static commandName = "make:listener";
	static description = "Create a new event listener class";
	static options = { allowUnknownFlags: true };
	stubPath = "make/listener/main.stub";
	prepare() {
		if (this.event) this.stubPath = "make/listener/for_event.stub";
	}
	async run() {
		const codemods = await this.createCodemods();
		if (this.event) {
			const { exitCode } = await this.kernel.exec("make:event", [this.event]);
			if (exitCode === 0) {
				const eventEntity = this.app.generators.createEntity(this.event);
				await codemods.makeUsingStub(stubsRoot, this.stubPath, {
					event: eventEntity,
					flags: this.parsed.flags,
					entity: this.app.generators.createEntity(this.name)
				}, { contentsFromFile: this.contentsFrom });
			}
			return;
		}
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the event listener" })], MakeListener.prototype, "name", void 0);
__decorate([flags.string({
	description: "Generate an event class alongside the listener",
	alias: "e"
})], MakeListener.prototype, "event", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeListener.prototype, "contentsFrom", void 0);
export { MakeListener as default };
