// @adonisjs/core/build/commands/make/middleware.js
import { t as stubsRoot } from "../../main-kn40V-hF.js";
import stringHelpers from "../../src/helpers/string.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
import string from "@poppinss/utils/string";
import { basename, extname, relative } from "node:path";
var MakeMiddleware = class extends BaseCommand {
	static commandName = "make:middleware";
	static description = "Create a new middleware class for HTTP requests";
	static options = { allowUnknownFlags: true };
	stubPath = "make/middleware/main.stub";
	async run() {
		const stackChoices = [
			"server",
			"router",
			"named"
		];
		if (!this.stack) this.stack = await this.prompt.choice("Under which stack you want to register the middleware?", stackChoices);
		if (!stackChoices.includes(this.stack)) {
			this.exitCode = 1;
			this.logger.error(`Invalid middleware stack "${this.stack}". Select from "${stackChoices.join(", ")}"`);
			return;
		}
		const codemods = await this.createCodemods();
		const { destination } = await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
		const middlewareRelativePath = stringHelpers.toUnixSlash(relative(this.app.middlewarePath(), destination).replace(extname(destination), ""));
		const name = string.camelCase(basename(middlewareRelativePath).replace(/_middleware$/, ""));
		await codemods.registerMiddleware(this.stack, [{
			name,
			path: `#middleware/${middlewareRelativePath}`
		}]);
	}
};
__decorate([args.string({ description: "Name of the middleware" })], MakeMiddleware.prototype, "name", void 0);
__decorate([flags.string({
	description: "The stack in which to register the middleware",
	alias: "s"
})], MakeMiddleware.prototype, "stack", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeMiddleware.prototype, "contentsFrom", void 0);
export { MakeMiddleware as default };
