// @adonisjs/core/build/commands/make/preload.js
import { t as stubsRoot } from "../../main-kn40V-hF.js";
import stringHelpers from "../../src/helpers/string.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
import { extname, relative } from "node:path";
const ALLOWED_ENVIRONMENTS = [
	"web",
	"console",
	"test",
	"repl"
];
var MakePreload = class extends BaseCommand {
	static commandName = "make:preload";
	static description = "Create a new preload file inside the start directory";
	stubPath = "make/preload/main.stub";
	#isEnvironmentsFlagValid() {
		if (!this.environments || !this.environments.length) return true;
		return this.environments.every((one) => ALLOWED_ENVIRONMENTS.includes(one));
	}
	async run() {
		if (!this.#isEnvironmentsFlagValid()) {
			this.logger.error(`Invalid environment(s) "${this.environments}". Only "${ALLOWED_ENVIRONMENTS}" are allowed`);
			return;
		}
		if (this.register === void 0) this.register = await this.prompt.confirm("Do you want to register the preload file in .adonisrc.ts file?");
		const codemods = await this.createCodemods();
		const { destination } = await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
		if (!this.register) return;
		const preloadFileRelativePath = stringHelpers.toUnixSlash(relative(this.app.startPath(), destination).replace(extname(destination), ""));
		await codemods.updateRcFile((rcFile) => {
			rcFile.addPreloadFile(`#start/${preloadFileRelativePath}`, this.environments);
		});
	}
};
__decorate([args.string({ description: "Name of the preload file" })], MakePreload.prototype, "name", void 0);
__decorate([flags.boolean({
	description: "Auto register the preload file inside the .adonisrc.ts file",
	showNegatedVariantInHelp: true,
	alias: "r"
})], MakePreload.prototype, "register", void 0);
__decorate([flags.array({
	description: `Define the preload file's environment. Accepted values are "${ALLOWED_ENVIRONMENTS}"`,
	alias: "e"
})], MakePreload.prototype, "environments", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakePreload.prototype, "contentsFrom", void 0);
export { MakePreload as default };
