// @adonisjs/core/build/commands/make/test.js
import { t as stubsRoot } from "../../main-kn40V-hF.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
var MakeTest = class extends BaseCommand {
	static commandName = "make:test";
	static description = "Create a new Japa test file";
	stubPath = "make/test/main.stub";
	async #getSuite() {
		if (this.suite) return this.suite;
		const rcFileSuites = this.app.rcFile.tests.suites;
		if (rcFileSuites.length === 1) return rcFileSuites[0].name;
		return this.prompt.choice("Select the suite for the test file", this.app.rcFile.tests.suites.map((suite) => {
			return suite.name;
		}), { validate(choice) {
			return choice ? true : "Please select a suite";
		} });
	}
	async #getSuiteDirectory(directories) {
		if (directories.length === 1) return directories[0];
		return this.prompt.choice("Select directory for the test file", directories, { validate(choice) {
			return choice ? true : "Please select a directory";
		} });
	}
	#findSuite(suiteName) {
		return this.app.rcFile.tests.suites.find((suite) => {
			return suite.name === suiteName;
		});
	}
	async run() {
		const suite = this.#findSuite(await this.#getSuite());
		if (!suite) {
			this.logger.error(`The "${this.suite}" suite is not configured inside the "adonisrc.js" file`);
			this.exitCode = 1;
			return;
		}
		await (await this.createCodemods()).makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name),
			suite: { directory: await this.#getSuiteDirectory(suite.directories) }
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the test file" })], MakeTest.prototype, "name", void 0);
__decorate([flags.string({
	description: "The suite for which to create the test file",
	alias: "s"
})], MakeTest.prototype, "suite", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeTest.prototype, "contentsFrom", void 0);
export { MakeTest as default };
