// @adonisjs/core/build/commands/make/validator.js
import { t as stubsRoot } from "../../main-kn40V-hF.js";
import { f as BaseCommand, l as flags, o as args } from "../../main-MBAMnmJb.js";
import { t as __decorate } from "../../decorate-DmrZA614.js";
var MakeValidator = class extends BaseCommand {
	static commandName = "make:validator";
	static description = "Create a new file to define VineJS validators";
	static options = { allowUnknownFlags: true };
	stubPath = "make/validator/main.stub";
	async prepare() {
		if (this.resource) this.stubPath = "make/validator/resource.stub";
	}
	async run() {
		await (await this.createCodemods()).makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the validator file" })], MakeValidator.prototype, "name", void 0);
__decorate([flags.boolean({ description: "Create a file with pre-defined validators for create and update actions" })], MakeValidator.prototype, "resource", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeValidator.prototype, "contentsFrom", void 0);
export { MakeValidator as default };
